// @kapso/whatsapp-cloud-api/dist/flow-json-BrA_xcig.cjs
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (all) => {
	let target = {};
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
	return target;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
let node_crypto = require("node:crypto");
node_crypto = __toESM(node_crypto);

//#region src/utils/case.ts
const camelCache = /* @__PURE__ */ new Map();
const snakeCache = /* @__PURE__ */ new Map();
function toCamelCaseKey(key) {
	if (!key.includes("_") && !key.includes("-")) return key;
	const cached = camelCache.get(key);
	if (cached) return cached;
	const result = key.replace(/[\-_]([a-z0-9])/gi, (_, char) => char.toUpperCase());
	camelCache.set(key, result);
	return result;
}
function toSnakeCaseKey(key) {
	if (!/[A-Z]/.test(key)) return key;
	const cached = snakeCache.get(key);
	if (cached) return cached;
	if (/^[A-Z0-9_]+$/.test(key)) {
		snakeCache.set(key, key);
		return key;
	}
	const result = key.replace(/([A-Z\d]+)([A-Z][a-z])/g, "$1_$2").replace(/([a-z\d])([A-Z])/g, "$1_$2").toLowerCase();
	snakeCache.set(key, result);
	return result;
}
function isPlainObject(value) {
	if (Object.prototype.toString.call(value) !== "[object Object]") return false;
	const proto = Object.getPrototypeOf(value);
	return proto === null || proto === Object.prototype;
}
function toCamelCaseDeep(input) {
	if (Array.isArray(input)) return input.map((item) => toCamelCaseDeep(item));
	if (isPlainObject(input)) {
		const result = {};
		for (const [key, value] of Object.entries(input)) result[toCamelCaseKey(key)] = toCamelCaseDeep(value);
		return result;
	}
	return input;
}
function toSnakeCaseDeep(input) {
	if (Array.isArray(input)) return input.map((item) => toSnakeCaseDeep(item));
	if (isPlainObject(input)) {
		const result = {};
		for (const [key, value] of Object.entries(input)) result[toSnakeCaseKey(key)] = toSnakeCaseDeep(value);
		return result;
	}
	return input;
}

//#endregion
//#region src/utils/flow-json.ts
const UNDERSCORE_KEYS = new Map([
	["dataApiVersion", "data_api_version"],
	["routingModel", "routing_model"],
	["refreshOnBack", "refresh_on_back"],
	["dataChannelUri", "data_channel_uri"],
	["previewParameters", "preview_parameters"],
	["validationErrors", "validation_errors"],
	["healthStatus", "health_status"],
	["jsonVersion", "json_version"],
	["dataApi", "data_api"],
	["endpointUri", "endpoint_uri"]
]);
const KEBAB_KEYS = new Map([
	["onClickAction", "on-click-action"],
	["onSelectAction", "on-select-action"],
	["onUnselectAction", "on-unselect-action"],
	["onDeselectAction", "on-deselect-action"],
	["onBackAction", "on-back-action"],
	["dataSource", "data-source"],
	["initValue", "init-value"],
	["initValues", "init-values"],
	["errorMessage", "error-message"],
	["helperText", "helper-text"],
	["minChars", "min-chars"],
	["maxChars", "max-chars"],
	["minLength", "min-length"],
	["maxLength", "max-length"],
	["labelVariant", "label-variant"],
	["fontWeight", "font-weight"],
	["maxUploadedPhotos", "max-uploaded-photos"],
	["maxUploadedDocuments", "max-uploaded-documents"],
	["minUploadedPhotos", "min-uploaded-photos"],
	["minUploadedDocuments", "min-uploaded-documents"],
	["photoSource", "photo-source"],
	["mediaSize", "media-size"],
	["altText", "alt-text"],
	["aspectRatio", "aspect-ratio"],
	["flowToken", "flow_token"],
	["flowAction", "flow_action"],
	["flowActionPayload", "flow_action_payload"],
	["phoneNumber", "phone_number"],
	["leftCaption", "left-caption"],
	["centerCaption", "center-caption"],
	["rightCaption", "right-caption"],
	["markdown", "markdown"],
	["limitedTimeOffer", "limited-time-offer"],
	["limitedTimeOfferExpirationTime", "limited-time-offer-expiration-time"],
	["listItems", "list-items"],
	["onClick", "on-click"],
	["onSelect", "on-select"],
	["onUnselect", "on-unselect"],
	["maxSelectedItems", "max-selected-items"],
	["minSelectedItems", "min-selected-items"],
	["descriptionText", "description-text"],
	["includeDays", "include-days"],
	["minDate", "min-date"],
	["maxDate", "max-date"],
	["unavailableDates", "unavailable-dates"],
	["mode", "mode"],
	["initError", "init-error"],
	["errorMessages", "error-messages"],
	["pageSize", "page-size"],
	["displayMode", "display-mode"],
	["capabilities", "capabilities"],
	["markdownEnabled", "markdown-enabled"],
	["mediaUrl", "media-url"],
	["flowActionData", "flow_action_data"],
	["flowId", "flow_id"],
	["flowName", "flow_name"]
]);
const REVERSE_UNDERSCORE_KEYS = /* @__PURE__ */ new Map();
const REVERSE_KEBAB_KEYS = /* @__PURE__ */ new Map();
for (const [camel, wire] of UNDERSCORE_KEYS.entries()) REVERSE_UNDERSCORE_KEYS.set(wire, camel);
for (const [camel, wire] of KEBAB_KEYS.entries()) REVERSE_KEBAB_KEYS.set(wire, camel);
const STRICT_IGNORE_KEYS = new Set(["__example__"]);
const PRESERVE_KEYS = new Set([
	"version",
	"screens",
	"id",
	"title",
	"terminal",
	"success",
	"sensitive",
	"data",
	"layout",
	"children",
	"type",
	"name",
	"visible",
	"required",
	"enabled",
	"form",
	"payload",
	"action",
	"screen",
	"flowToken",
	"flow_token",
	"flowActionPayload",
	"flow_action_payload",
	"flowAction",
	"flow_action",
	"flowId",
	"flow_id"
]);
let hashImplementation = initializeHashImplementation();
function initializeHashImplementation() {
	try {
		if (typeof node_crypto.createHash === "function") return (input) => (0, node_crypto.createHash)("sha256").update(input).digest("hex");
	} catch {
		return;
	}
}
function toFlowJsonWireCase(input, options = {}) {
	const { strictCamel = true } = options;
	const transform = (value, frame) => {
		if (Array.isArray(value)) return value.map((item) => transform(item, frame));
		if (isPlainObject(value)) {
			const isComponent = isComponentObject(value);
			const result = {};
			for (const [rawKey, rawVal] of Object.entries(value)) {
				if (strictCamel && shouldEnforceCamel(rawKey)) {
					const expected = suggestCamelCase(rawKey);
					throw new Error(`Flow JSON authoring should use camelCase key "${expected}" instead of "${rawKey}". Please update your source.`);
				}
				const mappedKey = mapAuthoringKey(rawKey, isComponent);
				result[mappedKey] = transform(rawVal, { isComponent });
			}
			return result;
		}
		return value;
	};
	return transform(input, { isComponent: false });
}
function fromFlowJsonWireCase(input) {
	const transform = (value) => {
		if (Array.isArray(value)) return value.map(transform);
		if (isPlainObject(value)) {
			const result = {};
			for (const [rawKey, rawVal] of Object.entries(value)) {
				const camelKey = mapWireKey(rawKey);
				result[camelKey] = transform(rawVal);
			}
			return result;
		}
		return value;
	};
	return transform(input);
}
function isComponentObject(value) {
	if (typeof value.type !== "string") return false;
	const type = value.type;
	return /^[A-Z]/.test(type);
}
function mapAuthoringKey(key, isComponent) {
	if (UNDERSCORE_KEYS.has(key)) return UNDERSCORE_KEYS.get(key);
	if (KEBAB_KEYS.has(key)) return KEBAB_KEYS.get(key);
	if (PRESERVE_KEYS.has(key)) return key;
	if (/^[A-Z0-9_]+$/.test(key)) return key;
	if (isComponent && /[A-Z]/.test(key)) return camelToKebab(key);
	return key;
}
function mapWireKey(key) {
	if (REVERSE_UNDERSCORE_KEYS.has(key)) return REVERSE_UNDERSCORE_KEYS.get(key);
	if (REVERSE_KEBAB_KEYS.has(key)) return REVERSE_KEBAB_KEYS.get(key);
	if (key.startsWith("__") && key.endsWith("__")) return key;
	if (key.includes("-")) return key.replace(/-([a-z])/g, (_, letter) => letter.toUpperCase());
	if (key.includes("_")) return key.replace(/_([a-z])/g, (_, letter) => letter.toUpperCase());
	return key;
}
function shouldEnforceCamel(key) {
	if (STRICT_IGNORE_KEYS.has(key)) return false;
	if (key.startsWith("__") && key.endsWith("__")) return false;
	return key.includes("_") || key.includes("-");
}
function suggestCamelCase(key) {
	return key.replace(/^__/, "__").replace(/[-_]+([a-z0-9])/gi, (_, char) => char.toUpperCase()).replace(/^([A-Z])/, (match) => match.toLowerCase());
}
function camelToKebab(key) {
	return key.replace(/([a-z0-9])([A-Z])/g, "$1-$2").replace(/([A-Z])([A-Z][a-z])/g, "$1-$2").toLowerCase();
}
function canonicalizeFlowJson(value) {
	const canonical = canonicalize(value);
	return JSON.stringify(canonical);
}
function canonicalize(value) {
	if (Array.isArray(value)) return value.map((item) => canonicalize(item));
	if (isPlainObject(value)) {
		const entries = Object.entries(value).sort(([a], [b]) => a < b ? -1 : a > b ? 1 : 0);
		const result = {};
		for (const [key, val] of entries) result[key] = canonicalize(val);
		return result;
	}
	return value;
}
function computeFlowJsonHash(value) {
	const canonical = canonicalizeFlowJson(value);
	if (hashImplementation) return hashImplementation(canonical);
	throw new Error("SHA-256 hashing is not available in this environment");
}
function fromWireKeyName(key) {
	return mapWireKey(key);
}

//#endregion
Object.defineProperty(exports, '__export', {
  enumerable: true,
  get: function () {
    return __export;
  }
});
Object.defineProperty(exports, '__toESM', {
  enumerable: true,
  get: function () {
    return __toESM;
  }
});
Object.defineProperty(exports, 'computeFlowJsonHash', {
  enumerable: true,
  get: function () {
    return computeFlowJsonHash;
  }
});
Object.defineProperty(exports, 'fromFlowJsonWireCase', {
  enumerable: true,
  get: function () {
    return fromFlowJsonWireCase;
  }
});
Object.defineProperty(exports, 'fromWireKeyName', {
  enumerable: true,
  get: function () {
    return fromWireKeyName;
  }
});
Object.defineProperty(exports, 'isPlainObject', {
  enumerable: true,
  get: function () {
    return isPlainObject;
  }
});
Object.defineProperty(exports, 'toCamelCaseDeep', {
  enumerable: true,
  get: function () {
    return toCamelCaseDeep;
  }
});
Object.defineProperty(exports, 'toFlowJsonWireCase', {
  enumerable: true,
  get: function () {
    return toFlowJsonWireCase;
  }
});
Object.defineProperty(exports, 'toSnakeCaseDeep', {
  enumerable: true,
  get: function () {
    return toSnakeCaseDeep;
  }
});
