// @kapso/whatsapp-cloud-api/dist/index.js
import { __export } from "./chunk-CTAAG5j7.js";
import { computeFlowJsonHash, fromWireKeyName, isPlainObject, toCamelCaseDeep, toFlowJsonWireCase, toSnakeCaseDeep } from "./flow-json-aEFnj21x.js";
import { z } from "zod";

//#region src/errors/categorize.ts
const codeCategoryMap = new Map([
	[0, "authorization"],
	[190, "authorization"],
	[3, "permission"],
	[10, "permission"],
	[200, "permission"],
	[201, "permission"],
	[202, "permission"],
	[203, "permission"],
	[204, "permission"],
	[205, "permission"],
	[206, "permission"],
	[207, "permission"],
	[208, "permission"],
	[209, "permission"],
	[210, "permission"],
	[211, "permission"],
	[212, "permission"],
	[213, "permission"],
	[214, "permission"],
	[215, "permission"],
	[216, "permission"],
	[217, "permission"],
	[218, "permission"],
	[219, "permission"],
	[4, "throttling"],
	[80007, "throttling"],
	[130429, "throttling"],
	[131048, "throttling"],
	[131056, "throttling"],
	[33, "parameter"],
	[100, "parameter"],
	[130472, "parameter"],
	[131008, "parameter"],
	[131009, "parameter"],
	[131021, "parameter"],
	[131026, "parameter"],
	[131051, "media"],
	[131052, "media"],
	[131053, "media"],
	[131e3, "server"],
	[131016, "server"],
	[131057, "server"],
	[133004, "server"],
	[133005, "server"],
	[368, "integrity"],
	[130497, "integrity"],
	[131031, "integrity"],
	[131047, "reengagementWindow"],
	[131037, "wabaConfig"],
	[131042, "businessEligibility"],
	[131045, "phoneRegistration"],
	[133e3, "phoneRegistration"],
	[133006, "phoneRegistration"],
	[133008, "phoneRegistration"],
	[133009, "phoneRegistration"],
	[133010, "phoneRegistration"],
	[133015, "phoneRegistration"],
	[133016, "phoneRegistration"],
	[132e3, "template"],
	[132001, "template"],
	[132005, "template"],
	[132007, "template"],
	[132012, "template"],
	[132015, "template"],
	[132016, "template"],
	[132068, "flow"],
	[132069, "flow"],
	[134011, "businessEligibility"],
	[135e3, "parameter"],
	[2593107, "synchronization"],
	[2593108, "synchronization"]
]);
const permissionRangeStart = 200;
const permissionRangeEnd = 299;
function categorizeErrorCode(code, httpStatus) {
	if (typeof code === "number") {
		const category = codeCategoryMap.get(code);
		if (category) return category;
		if (code >= permissionRangeStart && code <= permissionRangeEnd) return "permission";
	}
	if (httpStatus === 401) return "authorization";
	if (httpStatus === 403) return "permission";
	if (httpStatus === 404) return "parameter";
	if (httpStatus === 429) return "throttling";
	if (httpStatus >= 500) return "server";
	if (httpStatus >= 400 && httpStatus < 500) return "parameter";
	return "unknown";
}

//#endregion
//#region src/errors/retry.ts
const DO_NOT_RETRY_CODES = new Set([
	131049,
	131050,
	131047,
	368,
	130497,
	131031
]);
const REFRESH_TOKEN_CODES = new Set([0, 190]);
function deriveRetryHint({ category, code, httpStatus, retryAfterMs }) {
	if (typeof retryAfterMs === "number") return {
		action: "retry_after",
		retryAfterMs
	};
	if (typeof code === "number" && DO_NOT_RETRY_CODES.has(code)) return { action: "do_not_retry" };
	switch (category) {
		case "authorization": return { action: "refresh_token" };
		case "throttling": return { action: "retry_after" };
		case "integrity": return { action: "do_not_retry" };
		case "reengagementWindow": return { action: "do_not_retry" };
		case "businessEligibility":
		case "phoneRegistration":
		case "wabaConfig":
		case "template":
		case "media":
		case "permission":
		case "parameter":
		case "flow":
		case "synchronization": return { action: "fix_and_retry" };
		case "server": return httpStatus >= 500 ? { action: "retry" } : { action: "fix_and_retry" };
		case "unknown":
		default: return httpStatus >= 500 ? { action: "retry" } : { action: "fix_and_retry" };
	}
}
function shouldRefreshToken(code) {
	return typeof code === "number" && REFRESH_TOKEN_CODES.has(code);
}

//#endregion
//#region src/errors/graph-api-error.ts
var GraphApiError = class GraphApiError extends Error {
	httpStatus;
	code;
	type;
	details;
	errorSubcode;
	fbtraceId;
	errorData;
	category;
	retry;
	raw;
	constructor(params) {
		super(params.message);
		this.name = "GraphApiError";
		this.httpStatus = params.httpStatus;
		this.code = params.code;
		this.type = params.type;
		this.details = params.details;
		this.errorSubcode = params.errorSubcode;
		this.fbtraceId = params.fbtraceId;
		this.errorData = params.errorData;
		this.category = params.category;
		this.retry = params.retry;
		this.raw = params.raw;
	}
	static fromResponse(response, body, rawText) {
		const httpStatus = response.status;
		const retryAfterMs = parseRetryAfter(response.headers.get("Retry-After"));
		const camelBody = typeof body === "object" && body !== null ? toCamelCaseDeep(body) : body;
		if (isGraphErrorEnvelope(camelBody)) {
			const errorPayload = camelBody.error;
			const code = typeof errorPayload.code === "number" ? errorPayload.code : httpStatus;
			const type = errorPayload.type ?? GraphApiError.name;
			const details = typeof errorPayload.errorData?.details === "string" ? errorPayload.errorData.details : void 0;
			const category$1 = categorizeErrorCode(typeof errorPayload.code === "number" ? errorPayload.code : void 0, httpStatus);
			const retry$1 = deriveRetryHint({
				category: category$1,
				code: typeof errorPayload.code === "number" ? errorPayload.code : void 0,
				httpStatus,
				retryAfterMs
			});
			const message$1 = errorPayload.message ?? defaultMessage(httpStatus, details, rawText);
			return new GraphApiError({
				message: message$1,
				httpStatus,
				code,
				type,
				details,
				errorSubcode: errorPayload.errorSubcode,
				fbtraceId: errorPayload.fbtraceId,
				errorData: errorPayload.errorData,
				category: category$1,
				retry: retry$1,
				raw: camelBody
			});
		}
		if (camelBody && typeof camelBody === "object" && "error" in camelBody) {
			const { error } = camelBody;
			const errorMessage = typeof error === "string" ? error : defaultMessage(httpStatus, void 0, rawText);
			const category$1 = httpStatus >= 500 ? "server" : categorizeErrorCode(void 0, httpStatus);
			const retry$1 = deriveRetryHint({
				category: category$1,
				httpStatus,
				code: httpStatus,
				retryAfterMs
			});
			return new GraphApiError({
				message: errorMessage,
				httpStatus,
				code: httpStatus,
				type: GraphApiError.name,
				category: category$1,
				retry: retry$1,
				raw: camelBody
			});
		}
		const category = httpStatus >= 500 ? "server" : categorizeErrorCode(void 0, httpStatus);
		const retry = deriveRetryHint({
			category,
			httpStatus,
			code: httpStatus,
			retryAfterMs
		});
		const message = defaultMessage(httpStatus, void 0, rawText);
		return new GraphApiError({
			message,
			httpStatus,
			code: httpStatus,
			type: GraphApiError.name,
			category,
			retry,
			raw: rawText ?? body ?? null
		});
	}
	isAuthError() {
		return this.category === "authorization";
	}
	isRateLimit() {
		return this.category === "throttling";
	}
	isTemplateError() {
		return this.category === "template";
	}
	requiresTokenRefresh() {
		return this.category === "authorization" || shouldRefreshToken(typeof this.code === "number" ? this.code : void 0);
	}
	toJSON() {
		return {
			name: this.name,
			message: this.message,
			httpStatus: this.httpStatus,
			code: this.code,
			type: this.type,
			details: this.details,
			errorSubcode: this.errorSubcode,
			fbtraceId: this.fbtraceId,
			category: this.category,
			retry: this.retry,
			raw: this.raw
		};
	}
};
function isGraphErrorEnvelope(value) {
	if (!value || typeof value !== "object") return false;
	const payload = value.error;
	return typeof payload === "object" && payload !== null;
}
function parseRetryAfter(header) {
	if (!header) return;
	const numeric = Number(header);
	if (!Number.isNaN(numeric)) return Math.max(0, numeric) * 1e3;
	const date = Date.parse(header);
	if (!Number.isNaN(date)) {
		const diff = date - Date.now();
		return diff > 0 ? diff : 0;
	}
}
function defaultMessage(status, details, rawText) {
	if (details) return `Meta API request failed with status ${status}: ${details}`;
	if (rawText && rawText.trim().length > 0) return `Meta API request failed with status ${status}: ${rawText}`;
	return `Meta API request failed with status ${status}`;
}

//#endregion
//#region src/errors/kapso-proxy-required.ts
var KapsoProxyRequiredError = class extends Error {
	feature;
	helpUrl = "https://kapso.ai/";
	constructor(feature) {
		const message = `${feature} is only available via the Kapso Proxy. Set baseUrl to https://api.kapso.ai/meta/whatsapp and provide kapsoApiKey. Create a free account at https://kapso.ai/`;
		super(message);
		this.name = "KapsoProxyRequiredError";
		this.feature = feature;
	}
};

//#endregion
//#region src/resources/shared.ts
async function parseJsonResponse(response) {
	if (!response.ok) {
		const text = await safeReadText(response);
		let parsed;
		if (text) try {
			parsed = JSON.parse(text);
		} catch {
			parsed = void 0;
		}
		throw GraphApiError.fromResponse(response, parsed, text);
	}
	if ((response.headers.get("content-type") ?? "").includes("application/json")) {
		const json = await response.json();
		return toCamelCaseDeep(json);
	}
	if (response.status === 204) return null;
	return await response.text();
}
async function safeReadText(response) {
	try {
		return await response.text();
	} catch (error) {
		return String(error);
	}
}
/** Guard that ensures a Kapso-only feature is being used with the Kapso proxy base URL. */
function assertKapsoProxy(client, feature) {
	try {
		const c = client;
		if (!(typeof c?.isKapsoProxy === "function" ? c.isKapsoProxy() : inferKapsoFromBaseUrl(c?.baseUrl))) throw new KapsoProxyRequiredError(feature);
	} catch {
		throw new KapsoProxyRequiredError(feature);
	}
}
function inferKapsoFromBaseUrl(baseUrl) {
	if (!baseUrl) return false;
	try {
		return new URL(baseUrl).hostname.endsWith("kapso.ai");
	} catch {
		return false;
	}
}

//#endregion
//#region src/resources/messages/base.ts
const baseMessageSchema = z.object({
	phoneNumberId: z.string().min(1, "phoneNumberId is required"),
	to: z.string().min(1, "to is required"),
	recipientType: z.enum(["individual", "group"]).optional(),
	contextMessageId: z.string().min(1).optional(),
	bizOpaqueCallbackData: z.string().max(512).optional()
});
var MessageTransport = class {
	constructor(client) {
		this.client = client;
	}
	async send(phoneNumberId, payload) {
		return this.client.request("POST", `${phoneNumberId}/messages`, {
			body: payload,
			responseType: "json"
		});
	}
	async sendMessageRequest(phoneNumberId, payload) {
		return this.send(phoneNumberId, payload);
	}
};
function buildBasePayload(input, rest) {
	const payload = {
		messagingProduct: "whatsapp",
		recipientType: input.recipientType ?? "individual",
		to: input.to,
		...rest
	};
	if (input.contextMessageId) payload.context = { messageId: input.contextMessageId };
	if (input.bizOpaqueCallbackData) payload.bizOpaqueCallbackData = input.bizOpaqueCallbackData;
	return payload;
}

//#endregion
//#region src/resources/messages/text.ts
const textMessageSchema = baseMessageSchema.extend({
	body: z.string().min(1, "body must not be empty"),
	previewUrl: z.boolean().optional()
});
var TextMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = textMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "text",
			text: {
				body: parsed.body,
				...parsed.previewUrl !== void 0 ? { previewUrl: parsed.previewUrl } : {}
			}
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/image.ts
const mediaSchema = z.object({
	id: z.string().min(1).optional(),
	link: z.string().url().optional(),
	caption: z.string().max(1024).optional()
}).refine((value) => Boolean(value.id || value.link), { message: "Either id or link must be provided" });
const imageMessageSchema = baseMessageSchema.extend({ image: mediaSchema });
var ImageMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = imageMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "image",
			image: parsed.image
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/media.ts
const mediaRefBaseSchema = z.object({
	id: z.string().min(1).optional(),
	link: z.string().url().optional()
});
const withMediaRefine = (schema) => schema.refine((value) => {
	const candidate = value;
	return Boolean(candidate?.id || candidate?.link);
}, { message: "Either id or link must be provided" });
const mediaRefSchema = withMediaRefine(mediaRefBaseSchema);
const documentSchema = withMediaRefine(mediaRefBaseSchema.extend({
	caption: z.string().max(1024).optional(),
	filename: z.string().max(240).optional()
}));
const videoSchema = withMediaRefine(mediaRefBaseSchema.extend({ caption: z.string().max(1024).optional() }));
const audioSchema = withMediaRefine(mediaRefBaseSchema.extend({ voice: z.boolean().optional() }));
const stickerSchema = mediaRefSchema;
const documentMessageSchema = baseMessageSchema.extend({ document: documentSchema });
const videoMessageSchema = baseMessageSchema.extend({ video: videoSchema });
const audioMessageSchema = baseMessageSchema.extend({ audio: audioSchema });
const stickerMessageSchema = baseMessageSchema.extend({ sticker: stickerSchema });
var DocumentMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = documentMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "document",
			document: parsed.document
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};
var VideoMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = videoMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "video",
			video: parsed.video
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};
var AudioMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = audioMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "audio",
			audio: parsed.audio
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};
var StickerMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = stickerMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "sticker",
			sticker: parsed.sticker
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/location.ts
const locationSchema = z.object({
	latitude: z.number(),
	longitude: z.number(),
	name: z.string().max(100).optional(),
	address: z.string().max(300).optional()
});
const locationMessageSchema = baseMessageSchema.extend({ location: locationSchema });
var LocationMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = locationMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "location",
			location: parsed.location
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/contacts.ts
const addressSchema = z.object({
	street: z.string().optional(),
	city: z.string().optional(),
	state: z.string().optional(),
	zip: z.string().optional(),
	country: z.string().optional(),
	countryCode: z.string().optional(),
	type: z.string().optional()
});
const emailSchema = z.object({
	email: z.string().optional(),
	type: z.string().optional()
});
const phoneSchema = z.object({
	phone: z.string().optional(),
	waId: z.string().optional(),
	type: z.string().optional()
});
const urlSchema = z.object({
	url: z.string().optional(),
	type: z.string().optional()
});
const nameSchema = z.object({
	formattedName: z.string().min(1, "formattedName is required"),
	firstName: z.string().optional(),
	lastName: z.string().optional(),
	middleName: z.string().optional(),
	suffix: z.string().optional(),
	prefix: z.string().optional()
});
const orgSchema = z.object({
	company: z.string().optional(),
	department: z.string().optional(),
	title: z.string().optional()
});
const contactSchema = z.object({
	name: nameSchema,
	birthday: z.string().optional(),
	addresses: z.array(addressSchema).optional(),
	emails: z.array(emailSchema).optional(),
	org: orgSchema.optional(),
	phones: z.array(phoneSchema).optional(),
	urls: z.array(urlSchema).optional()
});
const contactsMessageSchema = baseMessageSchema.extend({ contacts: z.array(contactSchema).min(1, "At least one contact is required") });
var ContactsMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = contactsMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "contacts",
			contacts: parsed.contacts
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/template.ts
const templateComponentSchema = z.object({ type: z.string().min(1) }).loose();
const templateSchema = z.object({
	name: z.string().min(1, "template name is required"),
	language: z.object({
		code: z.string().min(1, "language code is required"),
		policy: z.literal("deterministic").optional()
	}),
	components: z.array(templateComponentSchema).optional()
}).transform((value) => value);
const templateMessageSchema = baseMessageSchema.extend({ template: templateSchema });
var TemplateMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = templateMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "template",
			template: parsed.template
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/reaction.ts
const reactionSchema = z.object({
	messageId: z.string().min(1, "messageId is required"),
	emoji: z.string().max(2).optional()
});
const reactionMessageSchema = baseMessageSchema.extend({ reaction: reactionSchema });
var ReactionMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	async send(input) {
		const parsed = reactionMessageSchema.parse(input);
		const payload = buildBasePayload(parsed, {
			type: "reaction",
			reaction: parsed.reaction
		});
		return this.client.sendMessageRequest(parsed.phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/interactive.ts
const MAX_BODY_CHARS = 1024;
const MAX_FOOTER_CHARS = 60;
const MAX_BUTTON_LABEL = 20;
const MAX_SECTION_TITLE = 24;
const MAX_CAROUSEL_CARD_BODY_CHARS = 160;
const MAX_CAROUSEL_CARD_BODY_LINE_BREAKS = 2;
function parseInput(schema, input, context) {
	const result = schema.safeParse(input);
	if (result.success) return result.data;
	const details = result.error.issues.map((issue) => {
		return `${issue.path.length ? issue.path.join(".") : "(root)"}: ${issue.message}`;
	}).join("; ");
	throw new Error(`${context} validation failed: ${details}`);
}
const textHeaderSchema = z.object({
	type: z.literal("text"),
	text: z.string().min(1).max(MAX_FOOTER_CHARS)
});
const headerMediaRef = z.object({
	id: z.string().min(1).optional(),
	link: z.string().url().optional()
}).refine((v) => Boolean(v.id || v.link), { message: "Either id or link must be provided" });
const imageHeaderSchema = z.object({
	type: z.literal("image"),
	image: headerMediaRef
});
const videoHeaderSchema = z.object({
	type: z.literal("video"),
	video: headerMediaRef
});
const documentHeaderSchema = z.object({
	type: z.literal("document"),
	document: headerMediaRef
});
const anyHeaderSchema = z.union([
	textHeaderSchema,
	imageHeaderSchema,
	videoHeaderSchema,
	documentHeaderSchema
]);
const buttonOptionSchema = z.object({
	id: z.string().min(1).max(256),
	title: z.string().min(1).max(MAX_BUTTON_LABEL)
});
const buttonMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	header: anyHeaderSchema.optional(),
	buttons: z.array(buttonOptionSchema).min(1).max(3)
});
const listRowSchema = z.object({
	id: z.string().min(1).max(200),
	title: z.string().min(1).max(MAX_SECTION_TITLE),
	description: z.string().max(72).optional()
});
const listSectionSchema = z.object({
	title: z.string().max(MAX_SECTION_TITLE).optional(),
	rows: z.array(listRowSchema).min(1).max(10)
});
const listMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(4096),
	buttonText: z.string().min(1).max(MAX_BUTTON_LABEL),
	sections: z.array(listSectionSchema).min(1).max(10),
	header: textHeaderSchema.optional(),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional()
});
const productMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS).optional(),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	header: textHeaderSchema.optional(),
	catalogId: z.string().min(1),
	productRetailerId: z.string().min(1)
});
const productItemSchema = z.object({ productRetailerId: z.string().min(1) });
const productListSectionSchema = z.object({
	title: z.string().min(1).max(MAX_SECTION_TITLE),
	productItems: z.array(productItemSchema).min(1).max(30)
});
const productListMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	header: textHeaderSchema.optional(),
	catalogId: z.string().min(1),
	sections: z.array(productListSectionSchema).min(1).max(10)
});
const flowActionPayloadSchema = z.object({}).catchall(z.unknown());
const addressSavedAddressSchema = z.object({
	id: z.string().min(1),
	value: z.object({
		name: z.string().optional(),
		phoneNumber: z.string().optional(),
		inPinCode: z.string().optional(),
		floorNumber: z.string().optional(),
		buildingName: z.string().optional(),
		address: z.string().optional(),
		landmarkArea: z.string().optional(),
		city: z.string().optional()
	}).catchall(z.unknown())
});
const addressParametersSchema = z.object({
	country: z.string().length(2),
	values: z.object({}).catchall(z.unknown()).optional(),
	savedAddresses: z.array(addressSavedAddressSchema).optional(),
	validationErrors: z.object({}).catchall(z.unknown()).optional()
});
const addressMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	parameters: addressParametersSchema
});
const locationRequestParametersSchema = z.object({ requestMessage: z.string().min(1).max(MAX_BODY_CHARS) });
const locationRequestMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	parameters: locationRequestParametersSchema
});
const callPermissionParametersSchema = z.object({
	phoneNumber: z.string().min(1),
	callPurpose: z.string().max(MAX_BODY_CHARS).optional()
});
const callPermissionMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().max(MAX_BODY_CHARS).optional(),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	parameters: callPermissionParametersSchema
});
const flowParametersSchema = z.object({
	flowId: z.string().min(1),
	flowCta: z.string().min(1).max(MAX_BUTTON_LABEL),
	mode: z.enum(["draft", "published"]).optional(),
	flowMessageVersion: z.string().min(1).optional(),
	flowToken: z.string().optional(),
	flowAction: z.enum(["navigate", "data_exchange"]).optional(),
	flowActionPayload: flowActionPayloadSchema.optional()
});
const flowMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	header: textHeaderSchema.optional(),
	parameters: flowParametersSchema
});
function buildHeader(header) {
	return header ? { ...header } : void 0;
}
function buildFooter(text) {
	return text ? { text } : void 0;
}
const ctaUrlParametersSchema = z.object({
	displayText: z.string().min(1).max(MAX_BUTTON_LABEL),
	url: z.string().url()
});
const ctaUrlMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS),
	header: anyHeaderSchema.optional(),
	footerText: z.string().max(MAX_FOOTER_CHARS).optional(),
	parameters: ctaUrlParametersSchema
});
const carouselCardBodyTextSchema = z.string().max(MAX_CAROUSEL_CARD_BODY_CHARS).refine((value) => (value.match(/\r\n|\r|\n/g) ?? []).length <= MAX_CAROUSEL_CARD_BODY_LINE_BREAKS, { message: `Must include at most ${MAX_CAROUSEL_CARD_BODY_LINE_BREAKS} line breaks` });
const carouselHeaderSchema = z.union([imageHeaderSchema, videoHeaderSchema]);
const carouselCtaActionSchema = z.object({
	type: z.literal("cta_url").optional(),
	displayText: z.string().min(1).max(MAX_BUTTON_LABEL),
	url: z.string().url()
});
const carouselQuickReplyActionSchema = z.object({ buttons: z.array(buttonOptionSchema).min(1) });
const carouselCardSchema$1 = z.object({
	cardIndex: z.number().int().min(0),
	header: carouselHeaderSchema,
	bodyText: carouselCardBodyTextSchema.optional(),
	action: z.union([carouselCtaActionSchema, carouselQuickReplyActionSchema])
});
const carouselMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS),
	cards: z.array(carouselCardSchema$1).min(2).max(10)
}).superRefine((value, ctx) => {
	const expectedIndexes = value.cards.map((_, index) => index);
	if (value.cards.map((card) => card.cardIndex).sort((left, right) => left - right).some((cardIndex, index) => cardIndex !== expectedIndexes[index])) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		path: ["cards"],
		message: "cardIndex values must be sequential from 0"
	});
	const firstAction = value.cards[0]?.action;
	const firstSignature = firstAction && "url" in firstAction ? "cta_url" : `quick_reply:${firstAction && "buttons" in firstAction ? firstAction.buttons.length : 0}`;
	const quickReplyIds = [];
	value.cards.forEach((card, index) => {
		if (("url" in card.action ? "cta_url" : `quick_reply:${card.action.buttons.length}`) !== firstSignature) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			path: [
				"cards",
				index,
				"action"
			],
			message: "All carousel cards must use the same button structure"
		});
		if ("buttons" in card.action) quickReplyIds.push(...card.action.buttons.map((button) => button.id));
	});
	if (new Set(quickReplyIds).size !== quickReplyIds.length) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		path: ["cards"],
		message: "quick reply button ids must be unique across carousel cards"
	});
});
const catalogMessageSchema = baseMessageSchema.extend({
	bodyText: z.string().min(1).max(MAX_BODY_CHARS).optional(),
	parameters: z.object({ thumbnailProductRetailerId: z.string().min(1).optional() }).default({})
});
var InteractiveMessageSender = class {
	constructor(client) {
		this.client = client;
	}
	assertTotalListRows(sections) {
		const total = sections.reduce((sum, s) => sum + (Array.isArray(s.rows) ? s.rows.length : 0), 0);
		if (total > 10) throw new Error(`List interactive message validation failed: total rows across all sections must be ≤ 10 (got ${total}).`);
	}
	async sendButtons(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, header, buttons } = parseInput(buttonMessageSchema, input, "Button interactive message");
		const interactive = {
			type: "button",
			body: { text: bodyText },
			action: { buttons: buttons.map((btn) => ({
				type: "reply",
				reply: {
					id: btn.id,
					title: btn.title
				}
			})) }
		};
		const builtHeader = buildHeader(header);
		if (builtHeader) Object.assign(interactive, { header: builtHeader });
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendList(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, buttonText, sections, header, footerText } = parseInput(listMessageSchema, input, "List interactive message");
		const interactive = {
			type: "list",
			body: { text: bodyText },
			action: {
				button: buttonText,
				sections: sections.map((section) => ({
					...section.title ? { title: section.title } : {},
					rows: section.rows.map((row) => ({
						id: row.id,
						title: row.title,
						...row.description ? { description: row.description } : {}
					}))
				}))
			}
		};
		this.assertTotalListRows(interactive.action.sections);
		const builtHeader = buildHeader(header);
		if (builtHeader) Object.assign(interactive, { header: builtHeader });
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendProduct(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, header, catalogId, productRetailerId } = parseInput(productMessageSchema, input, "Product interactive message");
		if (header) throw new Error("Product interactive message cannot include a header.");
		const interactive = {
			type: "product",
			action: {
				catalogId,
				productRetailerId
			}
		};
		if (bodyText) interactive.body = { text: bodyText };
		const builtHeader = buildHeader(header);
		if (builtHeader) Object.assign(interactive, { header: builtHeader });
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendProductList(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, header, catalogId, sections } = parseInput(productListMessageSchema, input, "Product list interactive message");
		if (!header) throw new Error("Product list interactive message requires a text header.");
		const totalProducts = sections.reduce((sum, section) => sum + section.productItems.length, 0);
		if (totalProducts > 30) throw new Error(`Product list interactive message validation failed: total products across sections must be ≤ 30 (got ${totalProducts}).`);
		const interactive = {
			type: "product_list",
			body: { text: bodyText },
			action: {
				catalogId,
				sections: sections.map((section) => ({
					title: section.title,
					productItems: section.productItems.map((item) => ({ productRetailerId: item.productRetailerId }))
				}))
			}
		};
		const builtHeader = buildHeader(header);
		if (builtHeader) Object.assign(interactive, { header: builtHeader });
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendFlow(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, header, parameters } = parseInput(flowMessageSchema, input, "Flow interactive message");
		const interactive = {
			type: "flow",
			body: { text: bodyText },
			action: {
				name: "flow",
				parameters: {
					flowCta: parameters.flowCta,
					flowId: parameters.flowId,
					flowMessageVersion: parameters.flowMessageVersion ?? "3",
					...parameters.mode ? { mode: parameters.mode } : {},
					...parameters.flowToken ? { flowToken: parameters.flowToken } : {},
					...parameters.flowAction ? { flowAction: parameters.flowAction } : {},
					...parameters.flowActionPayload ? { flowActionPayload: parameters.flowActionPayload } : {}
				}
			}
		};
		const builtHeader = buildHeader(header);
		if (builtHeader) Object.assign(interactive, { header: builtHeader });
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendAddress(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, parameters } = parseInput(addressMessageSchema, input, "Address interactive message");
		const interactive = {
			type: "address_message",
			body: { text: bodyText },
			action: {
				name: "address_message",
				parameters: {
					country: parameters.country,
					...parameters.values ? { values: parameters.values } : {},
					...parameters.savedAddresses ? { savedAddresses: parameters.savedAddresses } : {},
					...parameters.validationErrors ? { validationErrors: parameters.validationErrors } : {}
				}
			}
		};
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendLocationRequest(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, parameters } = parseInput(locationRequestMessageSchema, input, "Location request interactive message");
		const interactive = {
			type: "location_request_message",
			body: { text: bodyText },
			action: {
				name: "send_location",
				parameters: { requestMessage: parameters.requestMessage }
			}
		};
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendCallPermissionRequest(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, parameters } = parseInput(callPermissionMessageSchema, input, "Call permission interactive message");
		const interactive = {
			type: "call_permission_request",
			action: {
				name: "call_permission_request",
				parameters: {
					phoneNumber: parameters.phoneNumber,
					...parameters.callPurpose ? { callPurpose: parameters.callPurpose } : {}
				}
			}
		};
		if (bodyText) interactive.body = { text: bodyText };
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendCarousel(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, cards } = parseInput(carouselMessageSchema, input, "Carousel interactive message");
		const interactive = {
			type: "carousel",
			body: { text: bodyText },
			action: { cards: cards.map((card) => {
				const cardPayload = {
					cardIndex: card.cardIndex,
					type: "cta_url",
					header: card.header
				};
				if (card.bodyText) cardPayload.body = { text: card.bodyText };
				if ("url" in card.action) cardPayload.action = {
					name: "cta_url",
					parameters: {
						displayText: card.action.displayText,
						url: card.action.url
					}
				};
				else cardPayload.action = { buttons: card.action.buttons.map((button) => ({
					type: "quick_reply",
					quickReply: {
						id: button.id,
						title: button.title
					}
				})) };
				return cardPayload;
			}) }
		};
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendRaw(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, interactive } = input;
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendCtaUrl(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, footerText, header, parameters } = parseInput(ctaUrlMessageSchema, input, "CTA URL interactive message");
		const interactive = {
			type: "cta_url",
			body: { text: bodyText },
			action: {
				name: "cta_url",
				parameters: {
					displayText: parameters.displayText,
					url: parameters.url
				}
			}
		};
		const builtHeader = buildHeader(header);
		if (builtHeader) Object.assign(interactive, { header: builtHeader });
		const footer = buildFooter(footerText);
		if (footer) Object.assign(interactive, { footer });
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
	async sendCatalogMessage(input) {
		const { phoneNumberId, to, recipientType, contextMessageId, bizOpaqueCallbackData, bodyText, parameters } = parseInput(catalogMessageSchema, input, "Catalog interactive message");
		const interactive = {
			type: "catalog_message",
			action: {
				name: "catalog_message",
				parameters: { ...parameters.thumbnailProductRetailerId ? { thumbnailProductRetailerId: parameters.thumbnailProductRetailerId } : {} }
			}
		};
		if (bodyText) interactive.body = { text: bodyText };
		const payload = buildBasePayload({
			phoneNumberId,
			to,
			recipientType,
			contextMessageId,
			bizOpaqueCallbackData
		}, {
			type: "interactive",
			interactive
		});
		return this.client.sendMessageRequest(phoneNumberId, payload);
	}
};

//#endregion
//#region src/resources/messages/index.ts
const queryHistorySchema = z.object({
	phoneNumberId: z.string().min(1),
	direction: z.string().optional(),
	status: z.string().optional(),
	since: z.string().optional(),
	until: z.string().optional(),
	conversationId: z.string().optional(),
	limit: z.number().int().positive().max(100).optional(),
	after: z.string().optional(),
	before: z.string().optional(),
	fields: z.string().optional()
}).passthrough();
const getMessageSchema = z.object({
	phoneNumberId: z.string().min(1),
	messageId: z.string().min(1),
	fields: z.string().optional()
});
function cleanQuery$2(query) {
	return Object.fromEntries(Object.entries(query).filter(([, value]) => value !== void 0 && value !== null));
}
/**
* Send WhatsApp messages of all supported types.
*
* All methods return a typed SendMessageResponse.
* @category Messages
*/
var MessagesResource = class {
	transport;
	textSender;
	imageSender;
	audioSender;
	documentSender;
	videoSender;
	stickerSender;
	locationSender;
	contactsSender;
	templateSender;
	reactionSender;
	interactiveSender;
	constructor(client) {
		this.client = client;
		this.transport = new MessageTransport(client);
		this.textSender = new TextMessageSender(this.transport);
		this.imageSender = new ImageMessageSender(this.transport);
		this.audioSender = new AudioMessageSender(this.transport);
		this.documentSender = new DocumentMessageSender(this.transport);
		this.videoSender = new VideoMessageSender(this.transport);
		this.stickerSender = new StickerMessageSender(this.transport);
		this.locationSender = new LocationMessageSender(this.transport);
		this.contactsSender = new ContactsMessageSender(this.transport);
		this.templateSender = new TemplateMessageSender(this.transport);
		this.reactionSender = new ReactionMessageSender(this.transport);
		this.interactiveSender = new InteractiveMessageSender(this.transport);
	}
	async sendText(input) {
		return this.textSender.send(input);
	}
	async sendImage(input) {
		return this.imageSender.send(input);
	}
	async sendAudio(input) {
		return this.audioSender.send(input);
	}
	async sendDocument(input) {
		return this.documentSender.send(input);
	}
	async sendVideo(input) {
		return this.videoSender.send(input);
	}
	async sendSticker(input) {
		return this.stickerSender.send(input);
	}
	async sendLocation(input) {
		return this.locationSender.send(input);
	}
	async sendContacts(input) {
		return this.contactsSender.send(input);
	}
	async sendTemplate(input) {
		return this.templateSender.send(input);
	}
	async sendReaction(input) {
		return this.reactionSender.send(input);
	}
	async sendInteractiveButtons(input) {
		return this.interactiveSender.sendButtons(input);
	}
	async sendInteractiveList(input) {
		return this.interactiveSender.sendList(input);
	}
	async sendInteractiveProduct(input) {
		return this.interactiveSender.sendProduct(input);
	}
	async sendInteractiveProductList(input) {
		return this.interactiveSender.sendProductList(input);
	}
	async sendInteractiveFlow(input) {
		return this.interactiveSender.sendFlow(input);
	}
	async sendInteractiveAddress(input) {
		return this.interactiveSender.sendAddress(input);
	}
	async sendInteractiveLocationRequest(input) {
		return this.interactiveSender.sendLocationRequest(input);
	}
	async sendInteractiveCallPermission(input) {
		return this.interactiveSender.sendCallPermissionRequest(input);
	}
	async sendInteractiveCtaUrl(input) {
		return this.interactiveSender.sendCtaUrl(input);
	}
	async sendInteractiveCarousel(input) {
		return this.interactiveSender.sendCarousel(input);
	}
	async sendInteractiveCatalogMessage(input) {
		return this.interactiveSender.sendCatalogMessage(input);
	}
	async sendInteractiveRaw(input) {
		return this.interactiveSender.sendRaw(input);
	}
	async sendRaw(input) {
		return this.transport.send(input.phoneNumberId, input.payload);
	}
	async markRead(input) {
		const payload = {
			messagingProduct: "whatsapp",
			status: "read",
			messageId: input.messageId,
			...input.typingIndicator ? { typingIndicator: input.typingIndicator } : {}
		};
		return this.client.request("POST", `${input.phoneNumberId}/messages`, {
			body: payload,
			responseType: "json"
		});
	}
	async query(input) {
		assertKapsoProxy(this.client, "Message history API");
		const { phoneNumberId,...rest } = queryHistorySchema.parse(input);
		const query = cleanQuery$2(rest);
		return this.client.request("GET", `${phoneNumberId}/messages`, {
			query,
			responseType: "json"
		});
	}
	async get(input) {
		assertKapsoProxy(this.client, "Message history API");
		const { phoneNumberId, messageId, fields } = getMessageSchema.parse(input);
		const response = await this.client.request("GET", `${phoneNumberId}/messages/${messageId}`, {
			query: cleanQuery$2({ fields }),
			responseType: "json"
		});
		if (response && typeof response === "object" && "data" in response) return response.data;
		return response;
	}
	async listByConversation(input) {
		assertKapsoProxy(this.client, "Message history API");
		const { phoneNumberId, conversationId, limit, after, before, fields } = input;
		return this.query({
			phoneNumberId,
			conversationId,
			limit,
			after,
			before,
			fields
		});
	}
};

//#endregion
//#region src/resources/media.ts
const uploadSchema = z.object({
	phoneNumberId: z.string().min(1, "phoneNumberId is required"),
	type: z.string().min(1, "type is required"),
	file: z.custom(),
	fileName: z.string().min(1).optional(),
	messagingProduct: z.string().min(1).optional(),
	uploadStrategy: z.enum(["default", "resumable"]).optional()
});
const getSchema$3 = z.object({
	mediaId: z.string().min(1, "mediaId is required"),
	phoneNumberId: z.string().min(1).optional()
});
/**
* Manage media: upload, get metadata, and delete.
* @category Media
*/
var MediaResource = class {
	constructor(client) {
		this.client = client;
	}
	async upload(input) {
		const parsed = uploadSchema.parse(input);
		const form = new FormData();
		form.set("messaging_product", parsed.messagingProduct ?? "whatsapp");
		form.set("type", parsed.type);
		const blob = toBlob(parsed.file);
		if (parsed.fileName) form.set("file", blob, parsed.fileName);
		else form.set("file", blob);
		if (parsed.uploadStrategy) form.set("upload_strategy", parsed.uploadStrategy);
		return this.client.request("POST", `${parsed.phoneNumberId}/media`, {
			body: form,
			responseType: "json"
		});
	}
	async get(input) {
		const parsed = getSchema$3.parse(input);
		if (this.client.isKapsoProxy() && !parsed.phoneNumberId) throw new Error("phoneNumberId is required when using the Kapso proxy");
		return this.client.request("GET", parsed.mediaId, {
			query: buildMediaQuery(parsed.phoneNumberId),
			responseType: "json"
		});
	}
	async delete(input) {
		const parsed = getSchema$3.parse(input);
		if (this.client.isKapsoProxy() && !parsed.phoneNumberId) throw new Error("phoneNumberId is required when using the Kapso proxy");
		return this.client.request("DELETE", parsed.mediaId, {
			query: buildMediaQuery(parsed.phoneNumberId),
			responseType: "json"
		});
	}
	/**
	* Download media bytes for a given mediaId.
	* - Uses {@link get} to resolve the short-lived URL, then performs a raw fetch with client auth headers.
	* - Set `as` to control the return type. Defaults to `arrayBuffer`.
	*/
	async download(input) {
		const parsed = getSchema$3.parse(input);
		if (this.client.isKapsoProxy() && !parsed.phoneNumberId) throw new Error("phoneNumberId is required when using the Kapso proxy");
		const metadata = await this.get({
			mediaId: parsed.mediaId,
			phoneNumberId: parsed.phoneNumberId
		});
		const targetUrl = metadata.downloadUrl ?? metadata.url;
		const isKapsoHost = new URL(targetUrl).hostname.toLowerCase().endsWith("kapso.ai");
		const authMode = input.auth ?? "auto";
		const clientWithFetch = this.client;
		let res;
		if (authMode === "always" || authMode === "auto" && isKapsoHost) res = await clientWithFetch.fetch(targetUrl, { headers: input.headers });
		else {
			res = await clientWithFetch.rawFetch(targetUrl, { headers: input.headers });
			if ((res.status === 401 || res.status === 403) && authMode === "auto" && isKapsoHost) res = await clientWithFetch.fetch(targetUrl, { headers: input.headers });
		}
		if (input.as === "response") return res;
		if (input.as === "blob") return res.blob();
		return res.arrayBuffer();
	}
};
function buildMediaQuery(phoneNumberId) {
	if (!phoneNumberId) return;
	return { phoneNumberId };
}
function toBlob(uploadable) {
	if (uploadable instanceof Blob) return uploadable;
	if (typeof File !== "undefined" && uploadable instanceof File) return uploadable;
	if (uploadable instanceof ArrayBuffer) return new Blob([uploadable]);
	if (ArrayBuffer.isView(uploadable)) {
		const copy = new Uint8Array(uploadable.buffer, uploadable.byteOffset, uploadable.byteLength).slice();
		return new Blob([copy.buffer]);
	}
	throw new TypeError("Unsupported uploadable type");
}

//#endregion
//#region src/resources/templates/index.ts
const listSchema$3 = z.object({
	businessAccountId: z.string().min(1, "businessAccountId is required"),
	before: z.string().optional(),
	after: z.string().optional(),
	limit: z.number().int().positive().optional(),
	order: z.enum(["ASC", "DESC"]).optional(),
	status: z.string().optional(),
	name: z.string().optional(),
	category: z.string().optional(),
	language: z.string().optional()
});
const componentSchema = z.object({ type: z.string().min(1) }).loose();
const createSchema = z.object({
	businessAccountId: z.string().min(1, "businessAccountId is required"),
	name: z.string().min(1, "name is required"),
	language: z.string().min(1, "language is required"),
	category: z.string().min(1, "category is required"),
	parameterFormat: z.enum(["NAMED", "POSITIONAL"]).optional(),
	allowCategoryChange: z.boolean().optional(),
	components: z.array(componentSchema).nonempty(),
	qualityScoreCategory: z.string().optional()
});
const deleteSchema = z.object({
	businessAccountId: z.string().min(1, "businessAccountId is required"),
	name: z.string().min(1, "name is required"),
	language: z.string().optional()
});
const getSchema$2 = z.object({
	businessAccountId: z.string().min(1, "businessAccountId is required"),
	templateId: z.string().min(1, "templateId is required"),
	fields: z.string().optional()
});
/**
* CRUD for message templates on a WhatsApp Business Account (WABA).
* @category Templates
*/
var TemplatesResource = class {
	constructor(client) {
		this.client = client;
	}
	async list(input) {
		const { businessAccountId,...query } = listSchema$3.parse(input);
		return this.client.request("GET", `${businessAccountId}/message_templates`, {
			query: removeUndefined(query),
			responseType: "json"
		});
	}
	async create(input) {
		const { businessAccountId,...body } = createSchema.parse(input);
		return this.client.request("POST", `${businessAccountId}/message_templates`, {
			body,
			responseType: "json"
		});
	}
	async delete(input) {
		const { businessAccountId,...query } = deleteSchema.parse(input);
		return this.client.request("DELETE", `${businessAccountId}/message_templates`, {
			query: removeUndefined(query),
			responseType: "json"
		});
	}
	async get(input) {
		const { businessAccountId, templateId, fields } = getSchema$2.parse(input);
		return this.client.request("GET", `${businessAccountId}/message_templates/${templateId}`, {
			query: removeUndefined({ fields }),
			responseType: "json"
		});
	}
};
function removeUndefined(input) {
	const result = {};
	for (const [key, value] of Object.entries(input)) if (value !== void 0 && value !== null) result[key] = value;
	return result;
}

//#endregion
//#region src/resources/phone-numbers.ts
const requestCodeSchema = z.object({
	phoneNumberId: z.string().min(1),
	codeMethod: z.enum(["SMS", "VOICE"]),
	language: z.string().min(2)
});
const verifyCodeSchema = z.object({
	phoneNumberId: z.string().min(1),
	code: z.string().min(1)
});
const registerSchema = z.object({
	phoneNumberId: z.string().min(1),
	pin: z.string().min(1),
	dataLocalizationRegion: z.string().optional()
});
const deregisterSchema = z.object({ phoneNumberId: z.string().min(1) });
const settingsGetSchema = z.object({
	phoneNumberId: z.string().min(1),
	includeSipCredentials: z.boolean().optional()
});
const callingWeeklyHoursSchema = z.object({
	dayOfWeek: z.string().min(1),
	openTime: z.string().min(1),
	closeTime: z.string().min(1)
});
const callingHolidayScheduleSchema = z.object({
	date: z.string().min(1),
	startTime: z.string().min(1),
	endTime: z.string().min(1)
});
const callingHoursSchema = z.object({
	status: z.string().optional(),
	timezoneId: z.string().optional(),
	weeklyOperatingHours: z.array(callingWeeklyHoursSchema).optional(),
	holidaySchedule: z.array(callingHolidayScheduleSchema).optional()
});
const callingSipServerSchema = z.object({
	hostname: z.string().min(1),
	port: z.number().int().optional(),
	requestUriUserParams: z.record(z.string(), z.string()).optional(),
	sipUserPassword: z.string().optional()
});
const callingSipSchema = z.object({
	status: z.string().optional(),
	servers: z.array(callingSipServerSchema).optional()
});
const callingSettingsSchema = z.object({
	status: z.string().optional(),
	callIconVisibility: z.string().optional(),
	callHours: callingHoursSchema.optional(),
	callbackPermissionStatus: z.string().optional(),
	sip: callingSipSchema.optional()
});
const settingsUpdateSchema = z.object({
	phoneNumberId: z.string().min(1),
	fallbackLanguage: z.string().optional(),
	calling: callingSettingsSchema.optional()
}).passthrough();
const businessProfileGetSchema = z.object({
	phoneNumberId: z.string().min(1),
	fields: z.string().optional()
});
const businessProfileUpdateSchema = z.object({
	phoneNumberId: z.string().min(1),
	about: z.string().optional(),
	address: z.string().optional(),
	description: z.string().optional(),
	email: z.string().min(1).optional(),
	profilePictureUrl: z.string().url().optional(),
	websites: z.array(z.string().url()).optional(),
	vertical: z.string().optional()
}).refine((v) => Object.keys(v).length > 1, { message: "At least one field to update is required" });
/**
* Phone-number verification, registration, settings, and business profile.
* @category Phone Numbers
*/
var PhoneNumbersResource = class {
	constructor(client) {
		this.client = client;
	}
	async requestCode(input) {
		const parsed = requestCodeSchema.parse(input);
		return this.client.request("POST", `${parsed.phoneNumberId}/request_code`, {
			body: {
				codeMethod: parsed.codeMethod,
				language: parsed.language
			},
			responseType: "json"
		});
	}
	async verifyCode(input) {
		const parsed = verifyCodeSchema.parse(input);
		return this.client.request("POST", `${parsed.phoneNumberId}/verify_code`, {
			body: { code: parsed.code },
			responseType: "json"
		});
	}
	async register(input) {
		const parsed = registerSchema.parse(input);
		const body = {
			messagingProduct: "whatsapp",
			pin: parsed.pin
		};
		if (parsed.dataLocalizationRegion) body.dataLocalizationRegion = parsed.dataLocalizationRegion;
		return this.client.request("POST", `${parsed.phoneNumberId}/register`, {
			body,
			responseType: "json"
		});
	}
	async deregister(input) {
		const parsed = deregisterSchema.parse(input);
		return this.client.request("POST", `${parsed.phoneNumberId}/deregister`, { responseType: "json" });
	}
	settings = {
		get: async (input) => {
			const parsed = settingsGetSchema.parse(input);
			const query = parsed.includeSipCredentials ? { includeSipCredentials: true } : void 0;
			return this.client.request("GET", `${parsed.phoneNumberId}/settings`, {
				query,
				responseType: "json"
			});
		},
		update: async (input) => {
			const { phoneNumberId,...rest } = settingsUpdateSchema.parse(input);
			return this.client.request("POST", `${phoneNumberId}/settings`, {
				body: rest,
				responseType: "json"
			});
		}
	};
	businessProfile = {
		get: async (input) => {
			const { phoneNumberId, fields } = businessProfileGetSchema.parse(input);
			return this.client.request("GET", `${phoneNumberId}/whatsapp_business_profile`, {
				query: fields === void 0 ? void 0 : { fields },
				responseType: "json"
			});
		},
		update: async (input) => {
			const { phoneNumberId,...rest } = businessProfileUpdateSchema.parse(input);
			return this.client.request("POST", `${phoneNumberId}/whatsapp_business_profile`, {
				body: {
					messagingProduct: "whatsapp",
					...rest
				},
				responseType: "json"
			});
		}
	};
};

//#endregion
//#region src/resources/calls.ts
const sessionSchema = z.object({
	sdpType: z.string().min(1),
	sdp: z.string().min(1)
});
const connectSchema = z.object({
	phoneNumberId: z.string().min(1),
	to: z.string().min(1),
	session: sessionSchema.optional(),
	bizOpaqueCallbackData: z.string().max(512).optional()
});
const callIdSchema = z.object({
	phoneNumberId: z.string().min(1),
	callId: z.string().min(1)
});
const preAcceptSchema = callIdSchema.extend({ session: sessionSchema });
const acceptSchema = preAcceptSchema.extend({ bizOpaqueCallbackData: z.string().max(512).optional() });
const permissionsSchema = z.object({
	phoneNumberId: z.string().min(1),
	userWaId: z.string().min(1)
});
const listSchema$2 = z.object({
	phoneNumberId: z.string().min(1),
	direction: z.string().optional(),
	status: z.string().optional(),
	since: z.string().optional(),
	until: z.string().optional(),
	callId: z.string().optional(),
	limit: z.number().int().positive().max(100).optional(),
	after: z.string().optional(),
	before: z.string().optional(),
	fields: z.string().optional()
}).passthrough();
const getCallSchema = z.object({
	phoneNumberId: z.string().min(1),
	callId: z.string().min(1)
});
/**
* Calling API helpers: initiate/accept calls and manage permissions.
* @category Calls
*/
var CallsResource = class {
	constructor(client) {
		this.client = client;
	}
	async connect(input) {
		const { phoneNumberId, to, session, bizOpaqueCallbackData } = connectSchema.parse(input);
		const body = {
			messagingProduct: "whatsapp",
			to,
			action: "connect"
		};
		if (session) body.session = session;
		if (bizOpaqueCallbackData) body.bizOpaqueCallbackData = bizOpaqueCallbackData;
		return this.client.request("POST", `${phoneNumberId}/calls`, {
			body,
			responseType: "json"
		});
	}
	async preAccept(input) {
		const { phoneNumberId, callId, session } = preAcceptSchema.parse(input);
		return this.client.request("POST", `${phoneNumberId}/calls`, {
			body: {
				messagingProduct: "whatsapp",
				callId,
				action: "pre_accept",
				session
			},
			responseType: "json"
		});
	}
	async accept(input) {
		const { phoneNumberId, callId, session, bizOpaqueCallbackData } = acceptSchema.parse(input);
		const body = {
			messagingProduct: "whatsapp",
			callId,
			action: "accept",
			session
		};
		if (bizOpaqueCallbackData) body.bizOpaqueCallbackData = bizOpaqueCallbackData;
		return this.client.request("POST", `${phoneNumberId}/calls`, {
			body,
			responseType: "json"
		});
	}
	async reject(input) {
		const { phoneNumberId, callId } = callIdSchema.parse(input);
		return this.client.request("POST", `${phoneNumberId}/calls`, {
			body: {
				messagingProduct: "whatsapp",
				callId,
				action: "reject"
			},
			responseType: "json"
		});
	}
	async terminate(input) {
		const { phoneNumberId, callId } = callIdSchema.parse(input);
		return this.client.request("POST", `${phoneNumberId}/calls`, {
			body: {
				messagingProduct: "whatsapp",
				callId,
				action: "terminate"
			},
			responseType: "json"
		});
	}
	permissions = { get: async (input) => {
		const parsed = permissionsSchema.parse(input);
		return this.client.request("GET", `${parsed.phoneNumberId}/call_permissions`, {
			query: { userWaId: parsed.userWaId },
			responseType: "json"
		});
	} };
	async list(input) {
		assertKapsoProxy(this.client, "Calls history API");
		const { phoneNumberId,...rest } = listSchema$2.parse(input);
		const query = Object.fromEntries(Object.entries(rest).filter(([, value]) => value !== void 0 && value !== null));
		return this.client.request("GET", `${phoneNumberId}/calls`, {
			query,
			responseType: "json"
		});
	}
	async get(input) {
		assertKapsoProxy(this.client, "Calls history API");
		const { phoneNumberId, callId } = getCallSchema.parse(input);
		return (await this.list({
			phoneNumberId,
			callId,
			limit: 1
		})).data[0];
	}
};

//#endregion
//#region src/resources/conversations.ts
const listSchema$1 = z.object({
	phoneNumberId: z.string().min(1),
	status: z.string().min(1).optional(),
	lastActiveSince: z.string().min(1).optional(),
	lastActiveUntil: z.string().min(1).optional(),
	phoneNumber: z.string().min(1).optional(),
	limit: z.number().int().positive().max(100).optional(),
	after: z.string().optional(),
	before: z.string().optional(),
	fields: z.string().optional()
}).passthrough();
const getSchema$1 = z.object({ conversationId: z.string().min(1) });
const updateStatusSchema = z.object({
	conversationId: z.string().min(1),
	status: z.string().min(1)
});
function cleanQuery$1(query) {
	return Object.fromEntries(Object.entries(query).filter(([, value]) => value !== void 0 && value !== null));
}
/**
* Conversations history and lifecycle helpers.
* @category Conversations
*/
var ConversationsResource = class {
	constructor(client) {
		this.client = client;
	}
	async list(input) {
		assertKapsoProxy(this.client, "Conversations API");
		const { phoneNumberId,...rest } = listSchema$1.parse(input);
		const query = cleanQuery$1(rest);
		return this.client.request("GET", `${phoneNumberId}/conversations`, {
			query,
			responseType: "json"
		});
	}
	async get(input) {
		assertKapsoProxy(this.client, "Conversations API");
		const { conversationId } = getSchema$1.parse(input);
		const response = await this.client.request("GET", `conversations/${conversationId}`, { responseType: "json" });
		if (response && typeof response === "object" && "data" in response) return response.data;
		return response;
	}
	async updateStatus(input) {
		assertKapsoProxy(this.client, "Conversations API");
		const { conversationId, status } = updateStatusSchema.parse(input);
		return this.client.request("PATCH", `conversations/${conversationId}`, {
			body: { status },
			responseType: "json"
		});
	}
};

//#endregion
//#region src/resources/contacts.ts
const listSchema = z.object({
	phoneNumberId: z.string().min(1),
	waId: z.string().min(1).optional(),
	hasCustomer: z.boolean().optional(),
	customerId: z.string().uuid().optional(),
	limit: z.number().int().positive().max(100).optional(),
	after: z.string().optional(),
	before: z.string().optional(),
	fields: z.string().optional()
}).passthrough();
const getSchema = z.object({
	phoneNumberId: z.string().min(1),
	waId: z.string().min(1)
});
const updateSchema = z.object({
	phoneNumberId: z.string().min(1),
	waId: z.string().min(1),
	profileName: z.string().optional(),
	displayName: z.string().optional(),
	metadata: z.record(z.string(), z.unknown()).optional()
}).refine((value) => {
	return Object.keys(value).filter((key) => !["phoneNumberId", "waId"].includes(key)).length > 0;
}, { message: "At least one field to update is required" });
function cleanQuery(query) {
	return Object.fromEntries(Object.entries(query).filter(([, value]) => value !== void 0 && value !== null));
}
/**
* Contacts directory helpers.
* @category Contacts
*/
var ContactsResource = class {
	constructor(client) {
		this.client = client;
	}
	async list(input) {
		assertKapsoProxy(this.client, "Contacts API");
		const { phoneNumberId,...rest } = listSchema.parse(input);
		const query = cleanQuery(rest);
		return this.client.request("GET", `${phoneNumberId}/contacts`, {
			query,
			responseType: "json"
		});
	}
	async get(input) {
		assertKapsoProxy(this.client, "Contacts API");
		const { phoneNumberId, waId } = getSchema.parse(input);
		const response = await this.client.request("GET", `${phoneNumberId}/contacts/${waId}`, { responseType: "json" });
		if (response && typeof response === "object" && "data" in response) return response.data;
		return response;
	}
	async update(input) {
		assertKapsoProxy(this.client, "Contacts API");
		const { phoneNumberId, waId,...body } = updateSchema.parse(input);
		return this.client.request("PATCH", `${phoneNumberId}/contacts/${waId}`, {
			body,
			responseType: "json"
		});
	}
};

//#endregion
//#region src/resources/flows.ts
var FlowsResource = class {
	client;
	lastDeployedHashes = /* @__PURE__ */ new Map();
	constructor(client) {
		this.client = client;
	}
	/** Internal helper exposed for tests only. */
	__setLastDeployedHash(input) {
		const key = input.key ?? this.buildCacheKey(input.wabaId ?? "", input.name);
		this.lastDeployedHashes.set(key, { hash: input.hash });
	}
	async create(options) {
		const { wabaId, name, categories, flowJson, publish, endpointUri } = options;
		const body = {
			name,
			categories: categories && categories.length > 0 ? categories : ["OTHER"],
			flow_json: JSON.stringify(toFlowJsonWireCase(flowJson))
		};
		if (publish !== void 0) body.publish = publish;
		if (endpointUri) body.endpoint_uri = endpointUri;
		const response = await this.client.request("POST", `/${wabaId}/flows`, {
			body,
			responseType: "json"
		});
		return {
			id: response.id,
			success: response.success,
			validationErrors: normalizeValidationErrors(extractValidationErrors(response))
		};
	}
	async updateAsset(options) {
		const { flowId, json, file, phoneNumberId, businessAccountId } = options;
		if (!json && !file) throw new Error("Must supply either json or file when updating a Flow asset");
		const form = new FormData();
		if (json) {
			const payload = JSON.stringify(toFlowJsonWireCase(json));
			const blob = new Blob([payload], { type: "application/json" });
			form.append("file", blob, "flow.json");
		} else if (file instanceof Blob) form.append("file", file, "flow.json");
		else if (file instanceof ArrayBuffer || ArrayBuffer.isView(file)) {
			const buf = toStrictArrayBuffer(file);
			const blob = new Blob([buf], { type: "application/json" });
			form.append("file", blob, "flow.json");
		} else throw new Error("Unsupported file type for Flow asset upload");
		form.append("name", "flow.json");
		form.append("asset_type", "FLOW_JSON");
		const query = {};
		if (phoneNumberId) query.phoneNumberId = phoneNumberId;
		if (businessAccountId) query.businessAccountId = businessAccountId;
		const response = await this.client.request("POST", `/${flowId}/assets`, {
			body: form,
			query: Object.keys(query).length ? query : void 0,
			responseType: "json"
		});
		return {
			success: response.success,
			validationErrors: normalizeValidationErrors(extractValidationErrors(response))
		};
	}
	async publish(options) {
		const { flowId, phoneNumberId, businessAccountId } = options;
		const query = {};
		if (phoneNumberId) query.phoneNumberId = phoneNumberId;
		if (businessAccountId) query.businessAccountId = businessAccountId;
		return this.client.request("POST", `/${flowId}/publish`, {
			query: Object.keys(query).length ? query : void 0,
			responseType: "json"
		});
	}
	async deprecate(options) {
		const { flowId, phoneNumberId, businessAccountId } = options;
		const query = {};
		if (phoneNumberId) query.phoneNumberId = phoneNumberId;
		if (businessAccountId) query.businessAccountId = businessAccountId;
		return this.client.request("POST", `/${flowId}/deprecate`, {
			query: Object.keys(query).length ? query : void 0,
			responseType: "json"
		});
	}
	async preview(options) {
		const { flowId, interactive, fields, params, phoneNumberId, businessAccountId } = options;
		const queryParams = { fields: fields ?? (interactive ? "preview.invalidate(false)" : "preview") };
		if (params) Object.assign(queryParams, params);
		if (phoneNumberId) queryParams.phoneNumberId = phoneNumberId;
		if (businessAccountId) queryParams.businessAccountId = businessAccountId;
		const preview = (await this.client.request("GET", `/${flowId}`, {
			query: queryParams,
			responseType: "json"
		})).preview ?? {
			preview_url: "",
			expires_at: ""
		};
		return { preview: {
			previewUrl: preview.previewUrl ?? preview.preview_url ?? "",
			expiresAt: preview.expiresAt ?? preview.expires_at ?? ""
		} };
	}
	async get(options) {
		return this.client.request("GET", `/${options.flowId}`, {
			query: options.fields ? { fields: options.fields } : void 0,
			responseType: "json"
		});
	}
	async list(options) {
		const { wabaId, limit, after } = options;
		const query = {};
		if (limit !== void 0) query.limit = limit;
		if (after) query.after = after;
		return this.client.request("GET", `/${wabaId}/flows`, {
			query,
			responseType: "json"
		});
	}
	async deploy(flowJson, options) {
		const { name, wabaId, endpointUri, publish, preview, flowId: providedFlowId, categories, forceAssetUpload } = options;
		const cacheKey = this.buildCacheKey(wabaId, name);
		const hash = computeFlowJsonHash(flowJson);
		let flowId = providedFlowId;
		let validationErrors;
		if (!flowId) {
			const createResponse = await this.create({
				wabaId,
				name,
				categories,
				flowJson,
				endpointUri,
				publish: false
			});
			flowId = createResponse.id;
			validationErrors = createResponse.validationErrors;
			this.lastDeployedHashes.set(cacheKey, { hash });
		} else {
			const last = this.lastDeployedHashes.get(cacheKey);
			if (!last || last.hash !== hash || forceAssetUpload) {
				const update = await this.updateAsset({
					flowId,
					json: flowJson
				});
				if (update.validationErrors?.length) validationErrors = update.validationErrors;
				this.lastDeployedHashes.set(cacheKey, { hash });
			}
		}
		if (!flowId) throw new Error("Unable to resolve Flow ID after deployment");
		if (publish) await this.publish({ flowId });
		let previewUrl;
		if (preview) {
			const previewOptions = typeof preview === "boolean" ? { interactive: preview } : preview;
			previewUrl = (await this.preview({
				flowId,
				interactive: previewOptions.interactive,
				params: previewOptions.params
			})).preview.previewUrl || void 0;
		}
		return {
			flowId,
			previewUrl,
			validationErrors
		};
	}
	buildCacheKey(wabaId, name) {
		return `${wabaId}::${name}`;
	}
};
function toStrictArrayBuffer(input) {
	if (input instanceof ArrayBuffer) return input;
	const src = new Uint8Array(input.buffer, input.byteOffset, input.byteLength);
	const copy = new Uint8Array(src.byteLength);
	copy.set(src);
	return copy.buffer;
}
function normalizeValidationErrors(input) {
	if (!Array.isArray(input)) return void 0;
	const normalized = [];
	for (const candidate of input) {
		if (!isPlainObject(candidate)) continue;
		const errorCode = candidate.error;
		const entry = { error: typeof errorCode === "string" ? errorCode : String(errorCode ?? "") };
		for (const [key, value] of Object.entries(candidate)) {
			if (key === "error") continue;
			if (key === "error_type") {
				entry.errorType = typeof value === "string" ? value : String(value);
				continue;
			}
			if (key === "pointers" && Array.isArray(value)) {
				entry.pointers = value.filter(isPlainObject).map((pointer) => mapPointer(pointer));
				continue;
			}
			const camelKey = fromWireKeyName(key);
			entry[camelKey] = value;
		}
		const hint = deriveHint(entry.pointers);
		if (hint) entry.hint = hint;
		normalized.push(entry);
	}
	return normalized.length > 0 ? normalized : void 0;
}
function mapPointer(pointer) {
	const result = {};
	for (const [key, value] of Object.entries(pointer)) {
		if (key === "path") {
			result.path = typeof value === "string" ? value : String(value);
			continue;
		}
		const camelKey = fromWireKeyName(key);
		result[camelKey] = value;
	}
	return result;
}
function deriveHint(pointers) {
	if (!pointers) return void 0;
	for (const pointer of pointers) {
		const path = pointer.path;
		if (!path) continue;
		const match = path.match(/([A-Za-z0-9_-]+)(?!.*[A-Za-z0-9_-])/);
		if (!match) continue;
		const segment = match[1];
		if (!segment.includes("-") && !segment.includes("_")) continue;
		const camel = fromWireKeyName(segment);
		if (camel === segment) continue;
		return `Use ${camel} (camelCase). We map it to ${segment}.`;
	}
}
function extractValidationErrors(source) {
	if ("validation_errors" in source && source.validation_errors !== void 0) return source.validation_errors;
	if ("validationErrors" in source && source.validationErrors !== void 0) return source.validationErrors;
}

//#endregion
//#region src/client.ts
const DEFAULT_BASE_URL = "https://graph.facebook.com";
const DEFAULT_GRAPH_VERSION = "v23.0";
/**
* Minimal, fetch-based client for the WhatsApp Business Cloud API.
*
* - Supports calling Meta Graph directly or via Kapso proxy (set {@link WhatsAppClientConfig.baseUrl} and {@link WhatsAppClientConfig.kapsoApiKey}).
* - All resource helpers (messages, media, templates, phone-numbers) hang off this client.
* @category Client
*/
var WhatsAppClient = class {
	messages;
	media;
	templates;
	phoneNumbers;
	calls;
	conversations;
	contacts;
	flows;
	accessToken;
	kapsoApiKey;
	baseUrl;
	graphVersion;
	fetchImpl;
	kapsoProxy;
	constructor(config) {
		if (!config?.accessToken && !config?.kapsoApiKey) throw new Error("Must provide either an accessToken or kapsoApiKey");
		this.accessToken = config.accessToken;
		this.kapsoApiKey = config.kapsoApiKey;
		this.baseUrl = normalizeBaseUrl(config.baseUrl ?? DEFAULT_BASE_URL);
		this.graphVersion = config.graphVersion ?? DEFAULT_GRAPH_VERSION;
		this.kapsoProxy = detectKapsoProxy(this.baseUrl, this.kapsoApiKey);
		this.fetchImpl = config.fetch ?? globalThis.fetch;
		if (typeof this.fetchImpl !== "function") throw new Error("Global fetch API is not available. Please supply a custom fetch implementation.");
		this.messages = new MessagesResource(this);
		this.media = new MediaResource(this);
		this.templates = new TemplatesResource(this);
		this.phoneNumbers = new PhoneNumbersResource(this);
		this.calls = new CallsResource(this);
		this.conversations = new ConversationsResource(this);
		this.contacts = new ContactsResource(this);
		this.flows = new FlowsResource(this);
	}
	isKapsoProxy() {
		return this.kapsoProxy;
	}
	async request(method, path, options = {}) {
		const { responseType, query, headers: headerOverrides, body } = options;
		const url = this.buildUrl(path, query);
		const headers = this.buildHeaders(headerOverrides);
		const init = {
			method: method.toUpperCase(),
			headers
		};
		if (body !== void 0 && body !== null) if (isFormData(body) || isBlob(body) || body instanceof URLSearchParams || typeof body === "string" || body instanceof ArrayBuffer || ArrayBuffer.isView(body)) init.body = body;
		else {
			const payload = toSnakeCaseDeep(body);
			init.body = JSON.stringify(payload);
			if (!("Content-Type" in headers)) headers["Content-Type"] = "application/json";
		}
		const response = await this.fetchImpl(url, init);
		if (responseType === "json") return parseJsonResponse(response);
		return response;
	}
	/**
	* Perform a raw fetch to an absolute URL while automatically attaching the client's auth headers.
	* This does not modify the URL or payload (no base URL or version prefixing, no case conversion).
	*/
	async fetch(url, init = {}) {
		const { headers: headerOverrides,...rest } = init;
		const headers = this.buildHeaders(headerOverrides);
		const abs = typeof url === "string" ? url : url.toString();
		return this.fetchImpl(abs, {
			...rest,
			headers
		});
	}
	/**
	* Perform a raw fetch WITHOUT attaching auth headers.
	* Useful for downloading media from public CDNs that reject Authorization headers.
	*/
	async rawFetch(url, init = {}) {
		const abs = typeof url === "string" ? url : url.toString();
		return this.fetchImpl(abs, init);
	}
	buildHeaders(overrides) {
		const headers = {};
		if (this.accessToken) headers["Authorization"] = `Bearer ${this.accessToken}`;
		if (this.kapsoApiKey) headers["X-API-Key"] = this.kapsoApiKey;
		if (overrides) {
			for (const [key, value] of Object.entries(overrides)) if (value !== void 0 && value !== null) headers[key] = value;
		}
		return headers;
	}
	buildUrl(path, query) {
		const cleanedPath = stripLeadingSlash(path);
		const base = `${this.baseUrl}/${this.graphVersion}/`;
		const url = new URL(cleanedPath, ensureTrailingSlash(base));
		if (query) {
			const normalizedQuery = toSnakeCaseDeep(query);
			for (const [key, value] of Object.entries(normalizedQuery)) {
				if (value === void 0 || value === null) continue;
				if (Array.isArray(value)) {
					for (const item of value) if (item !== void 0 && item !== null) url.searchParams.append(key, String(item));
				} else url.searchParams.set(key, String(value));
			}
		}
		return url.toString();
	}
};
function stripLeadingSlash(input) {
	return input.replace(/^\/+/, "");
}
function ensureTrailingSlash(input) {
	return input.endsWith("/") ? input : `${input}/`;
}
function normalizeBaseUrl(input) {
	return input.replace(/\/+$/, "");
}
function detectKapsoProxy(baseUrl, kapsoApiKey) {
	if (typeof kapsoApiKey === "string" && kapsoApiKey.length > 0) return true;
	try {
		return new URL(baseUrl).hostname.endsWith("kapso.ai");
	} catch {
		return false;
	}
}
function isFormData(value) {
	return typeof FormData !== "undefined" && value instanceof FormData;
}
function isBlob(value) {
	return typeof Blob !== "undefined" && value instanceof Blob;
}

//#endregion
//#region src/resources/templates/definition.ts
var definition_exports = /* @__PURE__ */ __export({
	__testing: () => __testing,
	buildTemplateDefinition: () => buildTemplateDefinition
});
const NAME_REGEX = /^[a-z0-9_]{1,512}$/;
const LANGUAGE_REGEX = /^[a-z]{2}(?:_[A-Z]{2})?$/;
const CATEGORY_VALUES = [
	"UTILITY",
	"MARKETING",
	"AUTHENTICATION"
];
const HEADER_TEXT_MAX = 60;
const BODY_TEXT_MAX = 1024;
const FOOTER_TEXT_MAX = 60;
const CAROUSEL_BODY_TEXT_MAX = 160;
const OTP_SIGNATURE_REGEX = /^[A-Za-z0-9+/=]{11}$/;
const PACKAGE_NAME_REGEX = /^(?:[a-zA-Z][a-zA-Z0-9_]*\.)+[a-zA-Z][a-zA-Z0-9_]*$/;
const positionalPlaceholderRegex = /\{\{\s*(\d+)\s*\}\}/g;
const namedPlaceholderRegex = /\{\{\s*([a-zA-Z_][a-zA-Z0-9_]*)\s*\}\}/g;
function extractPlaceholders(text) {
	const positional = /* @__PURE__ */ new Set();
	const named = /* @__PURE__ */ new Set();
	text.replace(positionalPlaceholderRegex, (_, match) => {
		positional.add(match);
		return "";
	});
	text.replace(namedPlaceholderRegex, (_, match) => {
		if (!/^[0-9]+$/.test(match)) named.add(match);
		return "";
	});
	if (positional.size > 0 && named.size > 0) throw new Error("Cannot mix positional and named placeholders in the same string");
	return {
		positionalCount: positional.size,
		namedKeys: named
	};
}
const HeaderTextExampleSchema = z.object({ headerText: z.array(z.string().min(1)).min(1) }).or(z.object({ headerTextNamedParams: z.array(z.object({
	paramName: z.string().regex(/^[a-z0-9_]+$/),
	example: z.string().min(1)
})).min(1) }));
const BodyTextExampleSchema = z.object({ bodyText: z.array(z.array(z.string().min(1)).min(1)).min(1) }).or(z.object({ bodyTextNamedParams: z.array(z.object({
	paramName: z.string().regex(/^[a-z0-9_]+$/),
	example: z.string().min(1)
})).min(1) }));
const HeaderTextSchema = z.object({
	type: z.literal("HEADER"),
	format: z.literal("TEXT"),
	text: z.string().min(1).max(HEADER_TEXT_MAX),
	example: HeaderTextExampleSchema.optional()
}).superRefine((value, ctx) => {
	const { positionalCount, namedKeys } = extractPlaceholders(value.text);
	if (positionalCount === 0 && namedKeys.size === 0) {
		if (value.example) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "Header text example should be omitted when no placeholders are present",
			path: ["example"]
		});
		return;
	}
	if (!value.example) {
		ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "headerText example is required when placeholders are present",
			path: ["example"]
		});
		return;
	}
	if (positionalCount > 0) {
		if (!("headerText" in value.example)) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "headerText array must be provided for positional placeholders",
			path: ["example"]
		});
		else if (value.example.headerText.length !== positionalCount) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "headerText example count must match number of placeholders",
			path: ["example", "headerText"]
		});
	} else if (!("headerTextNamedParams" in value.example)) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "headerTextNamedParams must be provided for named placeholders",
		path: ["example"]
	});
	else {
		const provided = new Set(value.example.headerTextNamedParams.map((entry) => entry.paramName));
		for (const key of namedKeys) if (!provided.has(key)) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: `Missing example for named parameter ${key}`,
			path: ["example", "headerTextNamedParams"]
		});
	}
});
const HeaderMediaSchema = z.object({
	type: z.literal("HEADER"),
	format: z.enum([
		"IMAGE",
		"VIDEO",
		"DOCUMENT"
	]),
	example: z.object({ headerHandle: z.array(z.string().min(1)).min(1) })
});
const HeaderLocationSchema = z.object({
	type: z.literal("HEADER"),
	format: z.literal("LOCATION")
});
const HeaderComponentSchema = z.discriminatedUnion("format", [
	HeaderTextSchema,
	HeaderMediaSchema,
	HeaderLocationSchema,
	z.object({
		type: z.literal("HEADER"),
		format: z.literal("PRODUCT")
	})
]);
const BodyComponentSchema = z.object({
	type: z.literal("BODY"),
	text: z.string().min(1).max(BODY_TEXT_MAX).optional(),
	example: BodyTextExampleSchema.optional(),
	addSecurityRecommendation: z.boolean().optional()
}).superRefine((value, ctx) => {
	if (!value.text && !value.addSecurityRecommendation) {
		ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "BODY component must include text or addSecurityRecommendation",
			path: ["text"]
		});
		return;
	}
	if (!value.text) {
		if (value.example) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "Examples are not allowed when BODY text is omitted",
			path: ["example"]
		});
		return;
	}
	const { positionalCount, namedKeys } = extractPlaceholders(value.text);
	if (!(positionalCount > 0 || namedKeys.size > 0)) return;
	if (!value.example) {
		ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "bodyText example is required when placeholders are present",
			path: ["example"]
		});
		return;
	}
	if (positionalCount > 0) if (!("bodyText" in value.example)) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "bodyText example must be provided for positional placeholders",
		path: ["example"]
	});
	else {
		const firstRow = value.example.bodyText[0];
		if (!firstRow || firstRow.length !== positionalCount) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "bodyText example count must match number of placeholders",
			path: ["example", "bodyText"]
		});
	}
	else if (!("bodyTextNamedParams" in value.example)) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "bodyTextNamedParams must be provided for named placeholders",
		path: ["example"]
	});
	else {
		const provided = new Set(value.example.bodyTextNamedParams.map((entry) => entry.paramName));
		for (const key of namedKeys) if (!provided.has(key)) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: `Missing example for named parameter ${key}`,
			path: ["example", "bodyTextNamedParams"]
		});
	}
});
const FooterComponentSchema = z.object({
	type: z.literal("FOOTER"),
	text: z.string().min(1).max(FOOTER_TEXT_MAX).optional(),
	codeExpirationMinutes: z.number().int().min(1).max(90).optional()
}).superRefine((value, ctx) => {
	if (!value.text && value.codeExpirationMinutes === void 0) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "FOOTER must include text or codeExpirationMinutes",
		path: ["text"]
	});
	if (value.codeExpirationMinutes !== void 0 && value.codeExpirationMinutes > 10 && value.text) {}
});
const TemplateComponentSchema = z.union([
	HeaderComponentSchema,
	BodyComponentSchema,
	FooterComponentSchema,
	z.object({
		type: z.literal("BUTTONS"),
		buttons: z.array(z.discriminatedUnion("type", [
			z.object({
				type: z.literal("QUICK_REPLY"),
				text: z.string().min(1).max(25)
			}),
			z.object({
				type: z.literal("PHONE_NUMBER"),
				text: z.string().min(1).max(25),
				phoneNumber: z.string().min(1).max(20)
			}),
			z.object({
				type: z.literal("URL"),
				text: z.string().min(1).max(25),
				url: z.string().min(1).max(2e3),
				example: z.array(z.string().min(1)).min(1).max(1).optional()
			}).superRefine((v, ctx) => {
				if (/\{\{/.test(v.url) && !v.example) ctx.addIssue({
					code: z.ZodIssueCode.custom,
					message: "URL with variable requires example",
					path: ["example"]
				});
			}),
			z.object({
				type: z.literal("COPY_CODE"),
				example: z.string().min(1).max(15)
			}),
			z.object({
				type: z.literal("FLOW"),
				text: z.string().min(1).max(25).optional(),
				flowId: z.string().min(1).optional(),
				flowName: z.string().min(1).optional(),
				flowJson: z.string().min(1).optional(),
				flowAction: z.enum(["navigate", "data_exchange"]).optional(),
				navigateScreen: z.string().optional(),
				icon: z.enum([
					"DOCUMENT",
					"PROMOTION",
					"REVIEW"
				]).optional()
			}).superRefine((v, ctx) => {
				if ([
					v.flowId,
					v.flowName,
					v.flowJson
				].filter(Boolean).length !== 1) ctx.addIssue({
					code: z.ZodIssueCode.custom,
					message: "FLOW button requires exactly one of flowId, flowName, or flowJson"
				});
			}),
			z.object({
				type: z.literal("CATALOG"),
				text: z.string().min(1).max(25)
			}),
			z.object({
				type: z.literal("MPM"),
				text: z.string().min(1).max(25)
			}),
			z.object({
				type: z.literal("SPM"),
				text: z.string().min(1).max(25)
			}),
			z.object({
				type: z.literal("OTP"),
				otpType: z.enum([
					"COPY_CODE",
					"ONE_TAP",
					"ZERO_TAP"
				]),
				text: z.string().min(1).max(25).optional(),
				autofillText: z.string().min(1).max(25).optional(),
				supportedApps: z.array(z.object({
					packageName: z.string().regex(PACKAGE_NAME_REGEX, "Invalid package name"),
					signatureHash: z.string().regex(OTP_SIGNATURE_REGEX, "signatureHash must be 11 characters base64 string")
				})).min(1).max(5).optional(),
				packageName: z.string().regex(PACKAGE_NAME_REGEX, "Invalid package name").optional(),
				signatureHash: z.string().regex(OTP_SIGNATURE_REGEX, "signatureHash must be 11 characters base64 string").optional()
			}).superRefine((v, ctx) => {
				if (v.packageName && !v.signatureHash || !v.packageName && v.signatureHash) ctx.addIssue({
					code: z.ZodIssueCode.custom,
					message: "packageName and signatureHash must be provided together",
					path: v.packageName ? ["signatureHash"] : ["packageName"]
				});
				if (v.supportedApps) v.supportedApps.forEach((app, index) => {
					if (!OTP_SIGNATURE_REGEX.test(app.signatureHash)) ctx.addIssue({
						code: z.ZodIssueCode.custom,
						message: "supportedApps.signatureHash must be 11 characters base64 string",
						path: [
							"supportedApps",
							index,
							"signatureHash"
						]
					});
				});
			})
		])).min(1)
	}).superRefine((value, ctx) => {
		const btns = value.buttons;
		if (btns.length > 10) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "A maximum of 10 buttons is allowed",
			path: ["buttons"]
		});
		if (btns.filter((b) => b.type === "URL").length > 2) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "At most 2 URL buttons allowed",
			path: ["buttons"]
		});
		if (btns.filter((b) => b.type === "PHONE_NUMBER").length > 1) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "At most 1 PHONE_NUMBER button allowed",
			path: ["buttons"]
		});
		if (btns.filter((b) => b.type === "CATALOG").length > 1) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "At most 1 CATALOG button allowed",
			path: ["buttons"]
		});
		if (btns.filter((b) => b.type === "MPM").length > 1) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "At most 1 MPM button allowed",
			path: ["buttons"]
		});
		if (btns.filter((b) => b.type === "SPM").length > 1) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "At most 1 SPM button allowed",
			path: ["buttons"]
		});
		if (btns.filter((b) => b.type === "OTP").length > 1) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "At most 1 OTP button allowed",
			path: ["buttons"]
		});
		if (btns.some((b) => b.type === "OTP") && value.buttons.length > 1) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "OTP buttons cannot be combined with other buttons",
			path: ["buttons"]
		});
		const isQuick = (t) => t.type === "QUICK_REPLY";
		const groups = [];
		for (const b of btns) {
			const flag = isQuick(b);
			if (groups.length === 0 || groups[groups.length - 1] !== flag) groups.push(flag);
		}
		if (groups.length > 2) ctx.addIssue({
			code: z.ZodIssueCode.custom,
			message: "Quick reply buttons must be grouped, not interleaved",
			path: ["buttons"]
		});
	}),
	z.object({
		type: z.literal("LIMITED_TIME_OFFER"),
		limitedTimeOffer: z.object({
			text: z.string().min(1).max(16),
			hasExpiration: z.boolean().optional()
		})
	}),
	z.object({ type: z.literal("CALL_PERMISSION_REQUEST") }),
	z.object({
		type: z.literal("CAROUSEL"),
		cards: z.array(z.object({ components: z.array(z.discriminatedUnion("type", [
			z.discriminatedUnion("format", [z.object({
				type: z.literal("HEADER"),
				format: z.enum([
					"IMAGE",
					"VIDEO",
					"DOCUMENT"
				]),
				example: z.object({ headerHandle: z.array(z.string().min(1)).min(1) })
			}), z.object({
				type: z.literal("HEADER"),
				format: z.literal("PRODUCT")
			})]),
			z.object({
				type: z.literal("BODY"),
				text: z.string().min(1).max(CAROUSEL_BODY_TEXT_MAX),
				example: BodyTextExampleSchema.optional()
			}).superRefine((value, ctx) => {
				const { positionalCount, namedKeys } = extractPlaceholders(value.text);
				if (!(positionalCount > 0 || namedKeys.size > 0)) return;
				if (!value.example) {
					ctx.addIssue({
						code: z.ZodIssueCode.custom,
						message: "card body example is required when placeholders are present",
						path: ["example"]
					});
					return;
				}
				if (positionalCount > 0) if (!("bodyText" in value.example)) ctx.addIssue({
					code: z.ZodIssueCode.custom,
					message: "card bodyText example must be provided for positional placeholders",
					path: ["example"]
				});
				else {
					const firstRow = value.example.bodyText[0];
					if (!firstRow || firstRow.length !== positionalCount) ctx.addIssue({
						code: z.ZodIssueCode.custom,
						message: "card bodyText example count must match number of placeholders",
						path: ["example", "bodyText"]
					});
				}
				else if (!("bodyTextNamedParams" in value.example)) ctx.addIssue({
					code: z.ZodIssueCode.custom,
					message: "card bodyTextNamedParams must be provided for named placeholders",
					path: ["example"]
				});
				else {
					const provided = new Set(value.example.bodyTextNamedParams.map((entry) => entry.paramName));
					for (const key of namedKeys) if (!provided.has(key)) ctx.addIssue({
						code: z.ZodIssueCode.custom,
						message: `Missing example for named parameter ${key}`,
						path: ["example", "bodyTextNamedParams"]
					});
				}
			}),
			z.object({
				type: z.literal("BUTTONS"),
				buttons: z.array(z.discriminatedUnion("type", [
					z.object({
						type: z.literal("QUICK_REPLY"),
						text: z.string().min(1).max(25)
					}),
					z.object({
						type: z.literal("URL"),
						text: z.string().min(1).max(25),
						url: z.string().min(1).max(2e3),
						example: z.array(z.string().min(1)).min(1).max(1).optional()
					}).superRefine((v, ctx) => {
						if (/\{\{/.test(v.url) && !v.example) ctx.addIssue({
							code: z.ZodIssueCode.custom,
							message: "URL with variable requires example",
							path: ["example"]
						});
					}),
					z.object({
						type: z.literal("PHONE_NUMBER"),
						text: z.string().min(1).max(25),
						phoneNumber: z.string().min(1).max(20)
					}),
					z.object({
						type: z.literal("SPM"),
						text: z.string().min(1).max(25)
					})
				])).min(1).max(2)
			}).superRefine((value, ctx) => {
				if (value.buttons.filter((btn) => btn.type === "QUICK_REPLY").length > 1) ctx.addIssue({
					code: z.ZodIssueCode.custom,
					message: "At most 1 quick reply per carousel card",
					path: ["buttons"]
				});
			})
		])).min(1) })).min(2).max(10)
	})
]);
const TemplateDefinitionSchema = z.object({
	name: z.string().regex(NAME_REGEX, "Template name must be lowercase alphanumeric with underscores"),
	language: z.string().regex(LANGUAGE_REGEX, "Invalid language code"),
	category: z.enum(CATEGORY_VALUES),
	parameterFormat: z.enum(["NAMED", "POSITIONAL"]).optional(),
	messageSendTtlSeconds: z.number().int().positive().optional(),
	components: z.array(TemplateComponentSchema).nonempty()
}).superRefine((value, ctx) => {
	const bodyComponents = value.components.filter((component) => component.type === "BODY");
	if (bodyComponents.length === 0) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "Template must include a BODY component",
		path: ["components"]
	});
	else if (bodyComponents.length > 1) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "Template cannot include more than one BODY component",
		path: ["components"]
	});
	if (value.components.filter((component) => component.type === "HEADER").length > 1) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "Template cannot include more than one HEADER component",
		path: ["components"]
	});
	if (value.components.filter((component) => component.type === "FOOTER").length > 1) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "Template cannot include more than one FOOTER component",
		path: ["components"]
	});
	if (value.components.some((c) => c.type === "HEADER" && c.format === "LOCATION") && !["UTILITY", "MARKETING"].includes(value.category)) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "LOCATION headers are only allowed for UTILITY or MARKETING categories",
		path: ["components"]
	});
	if (value.components.find((c) => c.type === "LIMITED_TIME_OFFER") && value.category !== "MARKETING") ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "LIMITED_TIME_OFFER component allowed only for MARKETING category",
		path: ["components"]
	});
	if (value.components.find((c) => c.type === "CALL_PERMISSION_REQUEST") && !["MARKETING", "UTILITY"].includes(value.category)) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "CALL_PERMISSION_REQUEST component allowed only for MARKETING or UTILITY categories",
		path: ["components"]
	});
	if (value.components.filter((component) => component.type === "CAROUSEL").length > 1) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "Template cannot include more than one CAROUSEL component",
		path: ["components"]
	});
	if (value.components.some((component) => component.type === "BUTTONS" && component.buttons.some((btn) => btn.type === "OTP")) && value.category !== "AUTHENTICATION") ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "OTP buttons are only allowed for AUTHENTICATION templates",
		path: ["components"]
	});
});
function buildTemplateDefinition(input) {
	return TemplateDefinitionSchema.parse(input);
}
const __testing = {
	TemplateDefinitionSchema,
	HeaderComponentSchema,
	BodyComponentSchema
};

//#endregion
//#region src/resources/templates/send.ts
const headerParamSchema = z.discriminatedUnion("type", [
	z.object({
		type: z.literal("text"),
		text: z.string().min(1),
		parameterName: z.string().min(1).optional()
	}),
	z.object({
		type: z.literal("image"),
		image: z.object({
			id: z.string().min(1).optional(),
			link: z.string().url().optional()
		}).refine((v) => !!(v.id || v.link), { message: "image requires id or link" })
	}),
	z.object({
		type: z.literal("video"),
		video: z.object({
			id: z.string().min(1).optional(),
			link: z.string().url().optional()
		}).refine((v) => !!(v.id || v.link), { message: "video requires id or link" })
	}),
	z.object({
		type: z.literal("document"),
		document: z.object({
			id: z.string().min(1).optional(),
			link: z.string().url().optional()
		}).refine((v) => !!(v.id || v.link), { message: "document requires id or link" })
	}),
	z.object({
		type: z.literal("location"),
		location: z.object({
			latitude: z.union([z.number(), z.string()]),
			longitude: z.union([z.number(), z.string()]),
			name: z.string().min(1).optional(),
			address: z.string().min(1).optional()
		})
	})
]);
const bodyParamSchema = z.discriminatedUnion("type", [
	z.object({
		type: z.literal("text"),
		text: z.string(),
		parameterName: z.string().min(1).optional()
	}),
	z.object({
		type: z.literal("currency"),
		currency: z.object({
			fallbackValue: z.string(),
			code: z.string().min(1),
			amount1000: z.number().int()
		}),
		parameterName: z.string().min(1).optional()
	}),
	z.object({
		type: z.literal("date_time"),
		dateTime: z.object({ fallbackValue: z.string() }),
		parameterName: z.string().min(1).optional()
	})
]);
const buttonQuickReplySchema = z.object({
	type: z.literal("button"),
	subType: z.literal("quick_reply"),
	index: z.union([z.number().int().min(0).max(9), z.string()]),
	parameters: z.array(z.object({
		type: z.literal("payload"),
		payload: z.string().min(1),
		parameterName: z.string().min(1).optional()
	})).min(1)
});
const buttonUrlSchema = z.object({
	type: z.literal("button"),
	subType: z.literal("url"),
	index: z.union([z.number().int().min(0).max(9), z.string()]),
	parameters: z.array(z.object({
		type: z.literal("text"),
		text: z.string().min(1),
		parameterName: z.string().min(1).optional()
	})).min(1)
});
const buttonPhoneSchema = z.object({
	type: z.literal("button"),
	subType: z.literal("phone_number"),
	index: z.union([z.number().int().min(0).max(9), z.string()]),
	parameters: z.array(z.object({
		type: z.literal("text"),
		text: z.string().min(1),
		parameterName: z.string().min(1).optional()
	})).optional()
});
const buttonCopyCodeSchema = z.object({
	type: z.literal("button"),
	subType: z.literal("copy_code"),
	index: z.union([z.number().int().min(0).max(9), z.string()]),
	parameters: z.array(z.object({
		type: z.literal("coupon_code"),
		coupon_code: z.string().min(1).max(15)
	})).length(1)
});
const buttonFlowSchema = z.object({
	type: z.literal("button"),
	subType: z.literal("flow"),
	index: z.union([z.number().int().min(0).max(9), z.string()]),
	parameters: z.array(z.object({
		type: z.literal("action"),
		action: z.object({
			flow_token: z.string().min(1).optional(),
			flow_action_data: z.object({}).catchall(z.any()).optional()
		})
	})).optional()
});
const buttonParamSchema = z.union([
	buttonQuickReplySchema,
	buttonUrlSchema,
	buttonPhoneSchema,
	buttonCopyCodeSchema,
	buttonFlowSchema
]);
const carouselCardComponentSchema = z.union([
	z.object({
		type: z.literal("header"),
		parameters: z.tuple([headerParamSchema])
	}),
	z.object({
		type: z.literal("body"),
		parameters: z.array(bodyParamSchema).min(1)
	}),
	buttonParamSchema
]);
const carouselCardSchema = z.object({
	cardIndex: z.number().int().min(0),
	components: z.array(carouselCardComponentSchema).min(1)
});
const templateSendInputSchema = z.object({
	name: z.string().min(1),
	language: z.string().min(1),
	header: headerParamSchema.optional(),
	body: z.array(bodyParamSchema).optional(),
	cards: z.array(carouselCardSchema).min(1).optional(),
	buttons: z.array(buttonParamSchema).optional()
});
/**
* Build a Template message payload from typed parameters.
* Validates placeholders and structure with Zod before returning the payload expected by Meta.
* @category Templates
*/
function buildTemplateSendPayload(input) {
	if (Object.prototype.hasOwnProperty.call(input, "components")) throw new Error("buildTemplateSendPayload does not accept raw components; use buildTemplatePayload for Meta-style payloads");
	const parsed = templateSendInputSchema.parse(input);
	const components = [];
	if (parsed.header) components.push({
		type: "header",
		parameters: [parsed.header]
	});
	if (parsed.body && parsed.body.length > 0) components.push({
		type: "body",
		parameters: parsed.body
	});
	if (parsed.cards && parsed.cards.length > 0) components.push({
		type: "carousel",
		cards: parsed.cards
	});
	if (parsed.buttons && parsed.buttons.length > 0) for (const btn of parsed.buttons) components.push(btn);
	return {
		name: parsed.name,
		language: { code: parsed.language },
		components
	};
}

//#endregion
//#region src/kapso.ts
const KAPSO_MESSAGE_FIELDS = [
	"direction",
	"status",
	"processing_status",
	"phone_number",
	"has_media",
	"media_data",
	"media_url",
	"whatsapp_conversation_id",
	"contact_name",
	"message_type_data",
	"content",
	"flow_response",
	"flow_token",
	"flow_name",
	"order_text"
];
function buildKapsoFields(fields = KAPSO_MESSAGE_FIELDS) {
	const unique = Array.from(new Set(fields.map((field) => field.trim()).filter(Boolean)));
	if (unique.length === 0) return "kapso()";
	return `kapso(${unique.join(",")})`;
}
function buildKapsoMessageFields(...fields) {
	const flat = fields.flatMap((field) => Array.isArray(field) ? field : [field]);
	if (flat.length === 0) return buildKapsoFields();
	return buildKapsoFields(flat);
}

//#endregion
//#region src/resources/templates/raw.ts
const rawComponentSchema = z.object({
	type: z.string().min(1),
	parameters: z.any().optional(),
	sub_type: z.any().optional(),
	subType: z.any().optional(),
	index: z.any().optional()
}).passthrough();
const rawTemplatePayloadSchema = z.object({
	name: z.string().min(1),
	language: z.union([z.string().min(1), z.object({
		code: z.string().min(1),
		policy: z.literal("deterministic")
	}).strict()]),
	components: z.array(rawComponentSchema)
}).strict();
function buildTemplatePayload(input) {
	let parsed;
	try {
		parsed = rawTemplatePayloadSchema.parse(input);
	} catch (error) {
		if (error instanceof z.ZodError) {
			const typeIssue = error.issues.find((issue) => issue.path.length >= 3 && issue.path[0] === "components" && issue.path[2] === "type");
			if (typeIssue) {
				const componentIndex = typeIssue.path[1];
				throw new Error(`components[${String(componentIndex)}].type is required`);
			}
		}
		throw error;
	}
	const language = typeof parsed.language === "string" ? { code: parsed.language } : parsed.language;
	const components = parsed.components.map((component, componentIndex) => normalizeTemplateComponent(component, `components[${componentIndex}]`));
	return {
		name: parsed.name,
		language,
		components
	};
}
function normalizeTemplateComponent(component, path, options = { allowCarousel: true }) {
	const typeRaw = component.type;
	if (typeof typeRaw !== "string") throw new Error(`${path}.type is required`);
	switch (typeRaw.toLowerCase()) {
		case "header": return {
			type: "header",
			parameters: [normalizeHeaderParameter(component.parameters, path)]
		};
		case "body": return {
			type: "body",
			parameters: normalizeBodyParameters(component.parameters, path)
		};
		case "button": return normalizeButtonComponent(component, path);
		case "carousel":
			if (options.allowCarousel === false) throw new Error(`${path}.type 'carousel' is not supported inside carousel cards`);
			return normalizeCarouselComponent(component, path);
		default: throw new Error(`${path}.type '${String(component.type)}' is not supported`);
	}
}
function normalizeCarouselComponent(component, path) {
	const cards = component.cards;
	if (!Array.isArray(cards) || cards.length === 0) throw new Error(`${path}.cards must be a non-empty array`);
	return {
		type: "carousel",
		cards: cards.map((card, index) => normalizeCarouselCard(card, `${path}.cards[${index}]`))
	};
}
function normalizeCarouselCard(card, path) {
	if (!card || typeof card !== "object" || Array.isArray(card)) throw new Error(`${path} must be an object`);
	const cardRecord = card;
	const cardIndexRaw = cardRecord.cardIndex ?? cardRecord.card_index;
	if (cardIndexRaw === void 0 || cardIndexRaw === null) throw new Error(`${path}.card_index is required`);
	const cardIndex = typeof cardIndexRaw === "string" ? Number(cardIndexRaw) : cardIndexRaw;
	if (typeof cardIndex !== "number" || !Number.isInteger(cardIndex) || cardIndex < 0) throw new Error(`${path}.card_index must be a non-negative integer`);
	const components = cardRecord.components;
	if (!Array.isArray(components) || components.length === 0) throw new Error(`${path}.components must be a non-empty array`);
	return {
		cardIndex,
		components: components.map((component, index) => {
			if (!component || typeof component !== "object" || Array.isArray(component)) throw new Error(`${path}.components[${index}] must be an object`);
			const normalized = normalizeTemplateComponent(component, `${path}.components[${index}]`, { allowCarousel: false });
			if (normalized.type === "carousel") throw new Error(`${path}.components[${index}].type 'carousel' is not supported inside carousel cards`);
			return normalized;
		})
	};
}
function normalizeHeaderParameter(parameters, path) {
	if (!Array.isArray(parameters) || parameters.length !== 1) throw new Error(`${path}.parameters must be an array with exactly one entry for header components`);
	const param = clone(parameters[0]);
	const paramPath = `${path}.parameters[0]`;
	if (!param || typeof param !== "object" || typeof param.type !== "string") throw new Error(`${paramPath}.type is required`);
	const type = param.type.toLowerCase();
	param.type = type;
	if (type === "text") {
		const text = ensureString(param.text, `${paramPath}.text`);
		if (!text.trim()) throw new Error(`${paramPath}.text must be a non-empty string`);
		param.text = text;
		return param;
	}
	if (type === "image" || type === "video" || type === "document") {
		const mediaKey = type;
		const media = clone(param[mediaKey]);
		if (!media || typeof media !== "object") throw new Error(`${paramPath}.${mediaKey} must be an object containing id or link`);
		const id = media.id;
		const link = media.link;
		if (!id && !link) throw new Error(`${paramPath}.${mediaKey} requires id or link`);
		if (link) ensureValidUrl(link, `${paramPath}.${mediaKey}.link`);
		param[mediaKey] = media;
		return param;
	}
	if (type === "location") {
		const location = clone(param.location);
		if (!location || typeof location !== "object") throw new Error(`${paramPath}.location must be an object`);
		const latitude = location.latitude;
		const longitude = location.longitude;
		if (!isNumberLike(latitude) || !isNumberLike(longitude)) throw new Error(`${paramPath}.location.latitude and longitude must be provided`);
		param.location = {
			...location,
			latitude,
			longitude
		};
		return param;
	}
	throw new Error(`${paramPath}.type '${type}' is not supported in header component`);
}
function normalizeBodyParameters(parameters, path) {
	if (!Array.isArray(parameters) || parameters.length === 0) throw new Error(`${path}.parameters must be a non-empty array for body components`);
	return parameters.map((parameter, index) => {
		const paramPath = `${path}.parameters[${index}]`;
		const cloned = clone(parameter);
		if (!cloned || typeof cloned !== "object" || typeof cloned.type !== "string") throw new Error(`${paramPath}.type is required`);
		const type = cloned.type.toLowerCase();
		cloned.type = type;
		if (type === "text") {
			const text = ensureString(cloned.text, `${paramPath}.text`);
			if (!text.trim()) throw new Error(`${paramPath}.text must be a non-empty string`);
			cloned.text = text;
			return cloned;
		}
		if (type === "currency") {
			const currency = clone(cloned.currency);
			if (!currency || typeof currency !== "object") throw new Error(`${paramPath}.currency must be an object`);
			const fallbackValueKey = currency.fallbackValue !== void 0 ? "fallbackValue" : "fallback_value";
			ensureString(currency[fallbackValueKey], `${paramPath}.currency.fallback_value`);
			ensureString(currency.code, `${paramPath}.currency.code`);
			const amountKey = currency.amount1000 !== void 0 ? "amount1000" : "amount_1000";
			const amount = currency[amountKey];
			if (!Number.isInteger(amount)) throw new Error(`${paramPath}.currency.amount_1000 must be an integer`);
			cloned.currency = currency;
			return cloned;
		}
		if (type === "date_time") {
			const dateTimeKey = cloned.dateTime !== void 0 ? "dateTime" : "date_time";
			const dateTime = clone(cloned[dateTimeKey]);
			if (!dateTime || typeof dateTime !== "object") throw new Error(`${paramPath}.date_time must be an object`);
			const fallbackValueKey = dateTime.fallbackValue !== void 0 ? "fallbackValue" : "fallback_value";
			ensureString(dateTime[fallbackValueKey], `${paramPath}.date_time.fallback_value`);
			cloned[dateTimeKey] = dateTime;
			return cloned;
		}
		throw new Error(`${paramPath}.type '${type}' is not supported in body component`);
	});
}
function normalizeButtonComponent(component, path) {
	const subTypeRaw = component.subType ?? component.sub_type;
	const subType = ensureString(subTypeRaw, `${path}.sub_type`).toLowerCase();
	const index = component.index;
	if (index === void 0 || index === null) throw new Error(`${path}.index is required for button components`);
	const indexNumber = typeof index === "string" ? Number(index) : index;
	if (!Number.isInteger(indexNumber) || indexNumber < 0 || indexNumber > 9) throw new Error(`${path}.index must be an integer between 0 and 9`);
	const parameters = component.parameters;
	switch (subType) {
		case "quick_reply": {
			const params = normalizeQuickReplyParameters(parameters, path);
			return {
				type: "button",
				subType: "quick_reply",
				index: indexNumber,
				parameters: params
			};
		}
		case "url": {
			const params = normalizeButtonTextParameters(parameters, path);
			if (!params) throw new Error(`${path}.parameters must include at least one entry`);
			return {
				type: "button",
				subType: "url",
				index: indexNumber,
				parameters: params
			};
		}
		case "phone_number": {
			const params = normalizeButtonTextParameters(parameters, path, { optional: true });
			return params ? {
				type: "button",
				subType: "phone_number",
				index: indexNumber,
				parameters: params
			} : {
				type: "button",
				subType: "phone_number",
				index: indexNumber
			};
		}
		case "copy_code": {
			const params = normalizeButtonCopyCodeParameters(parameters, path);
			return {
				type: "button",
				subType: "copy_code",
				index: indexNumber,
				parameters: params
			};
		}
		case "flow": {
			const params = normalizeFlowParameters(parameters, path);
			return params ? {
				type: "button",
				subType: "flow",
				index: indexNumber,
				parameters: params
			} : {
				type: "button",
				subType: "flow",
				index: indexNumber
			};
		}
		case "catalog":
			if (parameters !== void 0 && !Array.isArray(parameters)) throw new Error(`${path}.parameters must be an array`);
			return parameters ? {
				type: "button",
				subType: "catalog",
				index: indexNumber,
				parameters
			} : {
				type: "button",
				subType: "catalog",
				index: indexNumber
			};
		default: throw new Error(`${path}.sub_type '${subType}' is not supported`);
	}
}
function ensureString(value, path) {
	if (typeof value !== "string") throw new Error(`${path} must be a string`);
	return value;
}
function ensureValidUrl(value, path) {
	if (typeof value !== "string") throw new Error(`${path} must be a string URL`);
	try {
		new URL(value);
	} catch {
		throw new Error(`${path} must be a valid URL`);
	}
}
function normalizeQuickReplyParameters(parameters, path) {
	if (!Array.isArray(parameters) || parameters.length === 0) throw new Error(`${path}.parameters must include at least one entry`);
	return parameters.map((param, paramIndex) => {
		const cloned = clone(param);
		if (!cloned || typeof cloned !== "object" || typeof cloned.type !== "string") throw new Error(`${path}.parameters[${paramIndex}].type is required`);
		cloned.type = cloned.type.toLowerCase();
		if (cloned.type !== "payload") throw new Error(`${path}.parameters[${paramIndex}].type must be 'payload'`);
		ensureString(cloned.payload, `${path}.parameters[${paramIndex}].payload`);
		return cloned;
	});
}
function normalizeButtonTextParameters(parameters, path, options = {}) {
	if (!Array.isArray(parameters) || parameters.length === 0) {
		if (options.optional) return;
		throw new Error(`${path}.parameters must include at least one entry`);
	}
	return parameters.map((param, index) => {
		const cloned = clone(param);
		if (!cloned || typeof cloned !== "object" || typeof cloned.type !== "string") throw new Error(`${path}.parameters[${index}].type is required`);
		cloned.type = cloned.type.toLowerCase();
		if (cloned.type !== "text") throw new Error(`${path}.parameters[${index}].type must be 'text'`);
		const text = ensureString(cloned.text, `${path}.parameters[${index}].text`);
		if (!text.trim()) throw new Error(`${path}.parameters[${index}].text must be a non-empty string`);
		if (options.maxLength && text.length > options.maxLength) throw new Error(`${path}.parameters[${index}].text must be <= ${options.maxLength} characters`);
		return cloned;
	});
}
function normalizeButtonCopyCodeParameters(parameters, path) {
	if (!Array.isArray(parameters) || parameters.length === 0) throw new Error(`${path}.parameters must include exactly one entry`);
	if (parameters.length !== 1) throw new Error(`${path}.parameters must include exactly one entry`);
	return parameters.map((param, index) => {
		const cloned = clone(param);
		if (!cloned || typeof cloned !== "object" || typeof cloned.type !== "string") throw new Error(`${path}.parameters[${index}].type is required`);
		cloned.type = cloned.type.toLowerCase();
		if (cloned.type !== "coupon_code") throw new Error(`${path}.parameters[${index}].type must be 'coupon_code'`);
		const code = ensureString(cloned.coupon_code, `${path}.parameters[${index}].coupon_code`);
		if (!code.trim()) throw new Error(`${path}.parameters[${index}].coupon_code must be a non-empty string`);
		if (code.length > 15) throw new Error(`${path}.parameters[${index}].coupon_code must be <= 15 characters`);
		return {
			type: "coupon_code",
			coupon_code: code
		};
	});
}
function normalizeFlowParameters(parameters, path) {
	if (parameters === void 0) return;
	if (!Array.isArray(parameters)) throw new Error(`${path}.parameters must be an array`);
	return parameters.map((param, index) => {
		const cloned = clone(param);
		if (cloned && typeof cloned === "object" && "type" in cloned && typeof cloned.type === "string") cloned.type = cloned.type.toLowerCase();
		return cloned;
	});
}
function clone(value) {
	if (value === null || typeof value !== "object") return value;
	return Array.isArray(value) ? value.map((item) => clone(item)) : { ...value };
}
function isNumberLike(value) {
	return typeof value === "number" || typeof value === "string" && value.trim() !== "" && !Number.isNaN(Number(value));
}

//#endregion
export { CallsResource, ContactsResource, ConversationsResource, FlowsResource, GraphApiError, KAPSO_MESSAGE_FIELDS, KapsoProxyRequiredError, MediaResource, MessagesResource, PhoneNumbersResource, definition_exports as TemplateDefinition, TemplatesResource, WhatsAppClient, buildKapsoFields, buildKapsoMessageFields, buildTemplatePayload, buildTemplateSendPayload, templateSendInputSchema };
