// @kapso/whatsapp-cloud-api/dist/server.js
import { fromFlowJsonWireCase, isPlainObject, toCamelCaseDeep } from "./flow-json-aEFnj21x.js";
import { createDecipheriv, createHash, createHmac, timingSafeEqual } from "node:crypto";

//#region src/webhooks/verify.ts
/**
* Verify X-Hub-Signature-256 for WhatsApp Webhooks using your app secret.
* Returns true when the signature matches the request body.
* @category Webhooks
*/
function verifySignature({ appSecret, rawBody, signatureHeader }) {
	try {
		if (!signatureHeader || typeof signatureHeader !== "string") return false;
		const [algo, received] = signatureHeader.split("=");
		if (algo !== "sha256" || !received) return false;
		const body = typeof rawBody === "string" ? Buffer.from(rawBody) : rawBody;
		const expected = createHmac("sha256", appSecret).update(body).digest("hex");
		const a = Buffer.from(received, "hex");
		const b = Buffer.from(expected, "hex");
		if (a.length !== b.length) return false;
		return timingSafeEqual(a, b);
	} catch {
		return false;
	}
}

//#endregion
//#region src/webhooks/normalize.ts
function normalizeWebhook(payload) {
	const result = {
		object: void 0,
		phoneNumberId: void 0,
		displayPhoneNumber: void 0,
		contacts: [],
		messages: [],
		statuses: [],
		calls: [],
		raw: {}
	};
	if (!payload || typeof payload !== "object") return result;
	const top = payload;
	if (typeof top.object === "string") result.object = top.object;
	const entries = Array.isArray(top.entry) ? top.entry : [];
	for (const entry of entries) {
		const changes = Array.isArray(entry?.changes) ? entry.changes : [];
		for (const change of changes) {
			const rawValue = change?.value;
			if (!rawValue || typeof rawValue !== "object") continue;
			const value = toCamelCaseDeep(rawValue);
			const fieldKey = toCamelField(change?.field);
			if (fieldKey) {
				const list = result.raw[fieldKey] ?? [];
				list.push(value);
				result.raw[fieldKey] = list;
			}
			const metadata = value.metadata ?? {};
			if (typeof metadata.phoneNumberId === "string") result.phoneNumberId = metadata.phoneNumberId;
			if (typeof metadata.displayPhoneNumber === "string") result.displayPhoneNumber = metadata.displayPhoneNumber;
			const contacts = Array.isArray(value.contacts) ? value.contacts : [];
			if (contacts.length > 0) for (const contact of contacts) result.contacts.push(toCamelCaseDeep(contact));
			const messages = Array.isArray(value.messages) ? value.messages : [];
			const messageEchoes = Array.isArray(value.messageEchoes) ? value.messageEchoes : [];
			for (const message of [...messages, ...messageEchoes]) {
				const normalized = normalizeMessage(message);
				applyDirection(normalized, metadata, messageEchoes.includes(message));
				result.messages.push(normalized);
			}
			const statuses = Array.isArray(value.statuses) ? value.statuses : [];
			for (const status of statuses) {
				const normalizedStatus = toCamelCaseDeep(status);
				result.statuses.push(normalizedStatus);
			}
			const calls = Array.isArray(value.calls) ? value.calls : [];
			for (const call of calls) {
				const normalizedCall = toCamelCaseDeep(call);
				if (typeof normalizedCall.wacid === "string" && !normalizedCall.callId) {
					normalizedCall.callId = normalizedCall.wacid;
					delete normalizedCall.wacid;
				}
				result.calls.push(normalizedCall);
			}
		}
	}
	return result;
}
function normalizeMessage(message) {
	const normalized = toCamelCaseDeep(message);
	if (normalized.order && typeof normalized.order.text !== "undefined") {
		const orderRecord = normalized.order;
		const orderText = orderRecord.text;
		delete orderRecord.text;
		if (orderText !== void 0) {
			orderRecord.orderText = orderText;
			ensureKapso(normalized).orderText = orderText;
		}
	}
	if (normalized.interactive?.type === "nfm_reply" && normalized.interactive.nfmReply && typeof normalized.interactive.nfmReply.responseJson === "string") {
		const nfm = normalized.interactive.nfmReply;
		const responseJson = nfm.responseJson;
		if (typeof responseJson === "string" && responseJson.trim()) try {
			const parsed = JSON.parse(responseJson);
			const camel = toCamelCaseDeep(parsed);
			const kapso = ensureKapso(normalized);
			kapso.flowResponse = camel;
			if (typeof camel.flowToken === "string") kapso.flowToken = camel.flowToken;
			if (typeof nfm.name === "string") kapso.flowName = nfm.name;
		} catch {}
	}
	if (normalized.kapso && Object.keys(normalized.kapso).length === 0) delete normalized.kapso;
	return normalized;
}
function applyDirection(message, metadata, isEcho) {
	const businessCandidates = [];
	const phoneNumberId = typeof metadata.phoneNumberId === "string" ? metadata.phoneNumberId : void 0;
	const displayPhoneNumber = typeof metadata.displayPhoneNumber === "string" ? metadata.displayPhoneNumber : void 0;
	const contextFrom = typeof message.context?.from === "string" ? message.context.from : void 0;
	if (phoneNumberId) businessCandidates.push(phoneNumberId);
	if (displayPhoneNumber) businessCandidates.push(displayPhoneNumber);
	if (contextFrom) businessCandidates.push(contextFrom);
	const businessSet = new Set(businessCandidates.map(normalizeNumber));
	const fromNorm = typeof message.from === "string" ? normalizeNumber(message.from) : void 0;
	const toNorm = typeof message.to === "string" ? normalizeNumber(message.to) : void 0;
	let direction;
	if (isEcho) direction = "outbound";
	else if (fromNorm && businessSet.has(fromNorm)) direction = "outbound";
	else if (toNorm && businessSet.has(toNorm)) direction = "inbound";
	else if (contextFrom && businessSet.has(normalizeNumber(contextFrom))) direction = "inbound";
	else if (fromNorm) direction = "inbound";
	if (direction || isEcho) {
		const kapso = ensureKapso(message);
		if (direction) kapso.direction = direction;
		if (isEcho) kapso.source = "smb_message_echo";
	}
}
function normalizeNumber(value) {
	if (!value) return "";
	return value.replace(/[^0-9]/g, "");
}
function ensureKapso(message) {
	if (!message.kapso) message.kapso = {};
	return message.kapso;
}
function toCamelField(field) {
	if (typeof field !== "string" || !field.length) return;
	return field.replace(/_([a-z])/g, (_, char) => char.toUpperCase());
}

//#endregion
//#region src/server/flows.ts
var FlowServerError = class extends Error {
	status;
	headers;
	body;
	constructor(status, message) {
		super(message);
		this.status = status;
		this.headers = { "Content-Type": "application/json" };
		this.body = JSON.stringify({ error: message });
	}
};
async function receive(options) {
	const text = Buffer.from(options.rawBody).toString("utf8");
	let body;
	try {
		body = JSON.parse(text);
	} catch {
		throw new FlowServerError(400, "Invalid JSON payload");
	}
	let decrypted = body;
	if (body.encrypted_flow_data && body.encryption_metadata) {
		const metadata = fromFlowJsonWireCase(body.encryption_metadata);
		decrypted = decryptFlowPayload(String(body.encrypted_flow_data), metadata);
	}
	const camel = fromFlowJsonWireCase(decrypted);
	const actionRaw = String(camel.action ?? "").toUpperCase();
	const action = actionRaw === "COMPLETE" ? "COMPLETE" : actionRaw === "BACK" ? "BACK" : "DATA_EXCHANGE";
	const screen = typeof camel.screen === "string" ? camel.screen : "";
	const flowToken = String(camel.flowToken ?? camel.flow_token ?? "");
	if (options.verifyToken) {
		if (!await options.verifyToken(flowToken, { phoneNumberId: options.phoneNumberId })) throw new FlowServerError(427, "Invalid flow token");
	}
	const form = isPlainObject(camel.form) ? fromFlowJsonWireCase(camel.form) : {};
	const data = isPlainObject(camel.data) ? fromFlowJsonWireCase(camel.data) : {};
	return {
		action,
		screen,
		flowToken,
		form,
		data,
		raw: camel
	};
}
function respond(options) {
	const status = options.status ?? 200;
	const headers = {
		"Content-Type": "application/json",
		...options.headers ?? {}
	};
	const body = JSON.stringify({
		screen: options.screen,
		data: options.data ?? {}
	});
	return {
		status,
		headers,
		body
	};
}
async function downloadAndDecrypt(options) {
	const fetchImpl = options.fetchImpl ?? globalThis.fetch;
	if (!fetchImpl) throw new Error("Global fetch API is not available. Provide fetchImpl explicitly.");
	const response = await fetchImpl(options.cdnUrl);
	if (!response.ok) throw new Error(`Failed to download media: ${response.status}`);
	const cipherWithTag = Buffer.from(await response.arrayBuffer());
	const plaintext = decryptBuffer(cipherWithTag, normalizeMetadata(options.encryptionMetadata));
	const copy = new Uint8Array(plaintext.length);
	copy.set(plaintext);
	return copy.buffer;
}
function decryptFlowPayload(encrypted, metadata) {
	const cipherWithTag = Buffer.from(encrypted, "base64");
	const meta = normalizeMetadata(metadata);
	const plaintext = decryptBuffer(cipherWithTag, meta);
	try {
		return JSON.parse(plaintext.toString("utf8"));
	} catch {
		throw new FlowServerError(400, "Unable to parse decrypted payload");
	}
}
function normalizeMetadata(metadata) {
	const encryptionKey = base64ToBuffer(String(metadata.encryptionKey ?? metadata.encryption_key ?? ""));
	const hmacKey = base64ToBuffer(String(metadata.hmacKey ?? metadata.hmac_key ?? ""));
	const iv = base64ToBuffer(String(metadata.iv ?? ""));
	const encryptedHash = base64ToBuffer(String(metadata.encryptedHash ?? metadata.encrypted_hash ?? ""));
	const plaintextHash = base64ToBuffer(String(metadata.plaintextHash ?? metadata.plaintext_hash ?? ""));
	if (encryptionKey.length === 0 || hmacKey.length === 0 || iv.length === 0) throw new FlowServerError(400, "Missing encryption metadata");
	return {
		encryptionKey,
		hmacKey,
		iv,
		encryptedHash,
		plaintextHash
	};
}
function decryptBuffer(cipherWithTag, metadata) {
	const encryptedHash = createHash("sha256").update(cipherWithTag).digest();
	if (!timingSafeEqual(encryptedHash, metadata.encryptedHash)) throw new FlowServerError(421, "Encrypted payload hash mismatch");
	if (cipherWithTag.length <= 10) throw new FlowServerError(421, "Invalid ciphertext length");
	const cipher = cipherWithTag.subarray(0, cipherWithTag.length - 10);
	const tag = cipherWithTag.subarray(cipherWithTag.length - 10);
	const computedHmac = createHmac("sha256", metadata.hmacKey).update(cipher).digest();
	if (!timingSafeEqual(tag, computedHmac.subarray(0, 10))) throw new FlowServerError(432, "HMAC validation failed");
	const decipher = createDecipheriv("aes-256-cbc", metadata.encryptionKey, metadata.iv);
	const unpadded = Buffer.concat([decipher.update(cipher), decipher.final()]);
	const plainHash = createHash("sha256").update(unpadded).digest();
	if (!timingSafeEqual(plainHash, metadata.plaintextHash)) throw new FlowServerError(421, "Plaintext hash mismatch");
	return unpadded;
}
function base64ToBuffer(value) {
	return Buffer.from(value, "base64");
}

//#endregion
export { FlowServerError, downloadAndDecrypt as downloadFlowMedia, normalizeWebhook, receive as receiveFlowEvent, respond as respondToFlow, verifySignature };
